{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "dAlJsXaqlfoo",
      "metadata": {
        "id": "dAlJsXaqlfoo"
      },
      "source": [
        "***Training course in data analysis for genomic surveillance of African malaria vectors - Workshop 1***\n",
        "\n",
        "---"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fe2e41ce-e1a7-4440-a10f-2ba36bceb543",
      "metadata": {
        "id": "fe2e41ce-e1a7-4440-a10f-2ba36bceb543"
      },
      "source": [
        "# Module 4 - Analysing SNPs in the *Vgsc* gene\n",
        "\n",
        "**Theme: Analysis**\n",
        "\n",
        "In this module we're going to perform an analysis to discover single nucleotide polymorphisms (SNPs) in the voltage-gated sodium channel gene (*Vgsc*), which encodes the binding target for pyrethroid insecticides. "
      ]
    },
    {
      "cell_type": "markdown",
      "id": "xDQA5mvAlo9Q",
      "metadata": {
        "id": "xDQA5mvAlo9Q"
      },
      "source": [
        "## Learning objectives\n",
        "\n",
        "At the end of this module you will be able to:\n",
        "\n",
        "* Discover mutations (SNPs) that could potentially cause pyrethroid target-site resistance.\n",
        "* Compute SNP allele frequencies, i.e., how common are they in different mosquito cohorts?\n",
        "* Perform analyses to compare SNP allele frequences between mosquitoes from different species, geographical locations and dates of collection.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "506c9e95-bb45-4248-a20a-0bf31057ab08",
      "metadata": {
        "id": "506c9e95-bb45-4248-a20a-0bf31057ab08"
      },
      "source": [
        "## Lecture"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cea426e6-3dee-487c-ad47-7524572525e4",
      "metadata": {
        "id": "cea426e6-3dee-487c-ad47-7524572525e4"
      },
      "source": [
        "### English"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "id": "df7cf784-6021-4cf6-9f2b-ca2850fe0be1",
      "metadata": {
        "tags": [
          "remove-input"
        ],
        "id": "df7cf784-6021-4cf6-9f2b-ca2850fe0be1",
        "outputId": "f8194be0-3763-40e2-baed-314241e820be",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 336
        }
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/m3R5SuNgKmw\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
            ],
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "%%html\n",
        "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/m3R5SuNgKmw\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ed68ea3e-de51-4303-b3ab-ad74e178d9c1",
      "metadata": {
        "id": "ed68ea3e-de51-4303-b3ab-ad74e178d9c1"
      },
      "source": [
        "### Fran√ßais\n",
        "\n",
        "Coming soon."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7df02bb6-4404-4a0c-a27c-db51068b98ff",
      "metadata": {
        "id": "7df02bb6-4404-4a0c-a27c-db51068b98ff"
      },
      "source": [
        "## Discovering SNPs in the *Vgsc* gene and computing SNP allele frequencies\n",
        "\n",
        "First, let's set up the [`malariagen_data`](https://github.com/malariagen/malariagen-data-python) package to access MalariaGEN data in the cloud."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "jliivGs3sPa3",
      "metadata": {
        "id": "jliivGs3sPa3",
        "tags": [
          "remove-output"
        ]
      },
      "outputs": [],
      "source": [
        "!pip install -q malariagen_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "vCKNT4xOsTvI",
      "metadata": {
        "id": "vCKNT4xOsTvI"
      },
      "outputs": [],
      "source": [
        "import malariagen_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "id": "n7f6J_CWsVtx",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 200
        },
        "id": "n7f6J_CWsVtx",
        "outputId": "7981d6b9-78c6-4b46-a310-8d93032a8eea"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "            <style type=\"text/css\">\n",
              "                table.malariagen-ag3 th, table.malariagen-ag3 td {\n",
              "                    text-align: left\n",
              "                }\n",
              "            </style>\n",
              "            <table class=\"malariagen-ag3\">\n",
              "                <thead>\n",
              "                    <tr>\n",
              "                        <th colspan=2>MalariaGEN Ag3 data resource API</th>\n",
              "                    </tr>\n",
              "                </thead>\n",
              "                <tbody>\n",
              "                    <tr><th>Storage URL</th><td>gs://vo_agam_release/</td></tr>\n",
              "                    <tr><th>Releases available</th><td>3.0</td></tr>\n",
              "                    <tr><th>Cohorts analysis</th><td>20211101</td></tr>\n",
              "                    <tr><th>Species analysis</th><td>aim_20200422</td></tr>\n",
              "                    <tr><th>Site filters analysis</th><td>dt_20200416</td></tr>\n",
              "                </tbody>\n",
              "            </table>\n",
              "            <p>Please note that data are subject to terms of use,\n",
              "            for more information see <a href=\"https://www.malariagen.net/data\">\n",
              "            the MalariaGEN website</a> or contact data@malariagen.net.</p>\n",
              "            <p>See also the <a href=\"https://malariagen.github.io/vector-data/ag3/api.html\">Ag3 API docs</a>.</p>\n",
              "        "
            ],
            "text/plain": [
              "<MalariaGEN Ag3 data resource API>\n",
              "Storage URL           : gs://vo_agam_release/\n",
              "Releases available    : 3.0\n",
              "Cohorts analysis      : 20211101\n",
              "Species analysis      : aim_20200422\n",
              "Site filters analysis : dt_20200416\n",
              "---\n",
              "Please note that data are subject to terms of use,\n",
              "for more information see https://www.malariagen.net/data\n",
              "or contact data@malariagen.net.\n",
              "---\n",
              "For API documentation see https://malariagen.github.io/vector-data/ag3/api.html"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ],
      "source": [
        "ag3 = malariagen_data.Ag3()\n",
        "ag3"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "Yr3xUejQsYKn",
      "metadata": {
        "id": "Yr3xUejQsYKn"
      },
      "source": [
        "To discover SNPs and compute allele frequencies, we're going to use the [`snp_allele_frequencies()` function](https://malariagen.github.io/vector-data/ag3/api.html#snp-allele-frequencies). Let's have a look at the documentation for this function."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "id": "_O5asOCZsqjP",
      "metadata": {
        "id": "_O5asOCZsqjP"
      },
      "outputs": [],
      "source": [
        "ag3.snp_allele_frequencies?"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "RBZlaBwzstG5",
      "metadata": {
        "id": "RBZlaBwzstG5"
      },
      "source": [
        "To discover SNPs in the *Vgsc* gene, we need to define some parameters. \n",
        "\n",
        "First, we need to decide which gene transcript to use when determining what SNP effects will be. Here we'll use the transcript with identifier \"AGAP004707-RD\"."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "id": "waElhaJNth8d",
      "metadata": {
        "id": "waElhaJNth8d"
      },
      "outputs": [],
      "source": [
        "transcript = \"AGAP004707-RD\""
      ]
    },
    {
      "cell_type": "markdown",
      "id": "T19_hRn-tmgo",
      "metadata": {
        "id": "T19_hRn-tmgo"
      },
      "source": [
        "Next, to compute allele frequencies, we need to decide how our mosquitoes will be grouped into cohorts. There are different ways you can do this, for this analysis we'll group spatially by level 1 administrative divisions within countries, and temporally by year."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "id": "2ISehnkwt0lo",
      "metadata": {
        "id": "2ISehnkwt0lo"
      },
      "outputs": [],
      "source": [
        "cohorts = \"admin1_year\""
      ]
    },
    {
      "cell_type": "markdown",
      "id": "Nlzv-dRRt2YB",
      "metadata": {
        "id": "Nlzv-dRRt2YB"
      },
      "source": [
        "Next, we need to choose which samples to analyse. There are a number of different sample sets in the Ag3.0 data resource that we could use for this analysis. Let's check what's available."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "id": "0lP6E82kuEkI",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 927
        },
        "id": "0lP6E82kuEkI",
        "outputId": "262687af-9824-4612-afd8-8a5659016449"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-65869631-83eb-44aa-be37-ddfd60cf32f6\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample_set</th>\n",
              "      <th>sample_count</th>\n",
              "      <th>release</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>AG1000G-AO</td>\n",
              "      <td>81</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>AG1000G-BF-A</td>\n",
              "      <td>181</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>AG1000G-BF-B</td>\n",
              "      <td>102</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>AG1000G-BF-C</td>\n",
              "      <td>13</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>AG1000G-CD</td>\n",
              "      <td>76</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>AG1000G-CF</td>\n",
              "      <td>73</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>AG1000G-CI</td>\n",
              "      <td>80</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>AG1000G-CM-A</td>\n",
              "      <td>303</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>AG1000G-CM-B</td>\n",
              "      <td>97</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>AG1000G-CM-C</td>\n",
              "      <td>44</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>AG1000G-FR</td>\n",
              "      <td>23</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>AG1000G-GA-A</td>\n",
              "      <td>69</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>AG1000G-GH</td>\n",
              "      <td>100</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>AG1000G-GM-A</td>\n",
              "      <td>74</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>AG1000G-GM-B</td>\n",
              "      <td>31</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>AG1000G-GM-C</td>\n",
              "      <td>174</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>AG1000G-GN-A</td>\n",
              "      <td>45</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>AG1000G-GN-B</td>\n",
              "      <td>185</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>AG1000G-GQ</td>\n",
              "      <td>10</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>AG1000G-GW</td>\n",
              "      <td>101</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>AG1000G-KE</td>\n",
              "      <td>86</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>AG1000G-ML-A</td>\n",
              "      <td>60</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>AG1000G-ML-B</td>\n",
              "      <td>71</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>AG1000G-MW</td>\n",
              "      <td>41</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>AG1000G-MZ</td>\n",
              "      <td>74</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>AG1000G-TZ</td>\n",
              "      <td>300</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>AG1000G-UG</td>\n",
              "      <td>290</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>AG1000G-X</td>\n",
              "      <td>297</td>\n",
              "      <td>3.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-65869631-83eb-44aa-be37-ddfd60cf32f6')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-65869631-83eb-44aa-be37-ddfd60cf32f6 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-65869631-83eb-44aa-be37-ddfd60cf32f6');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "      sample_set  sample_count release\n",
              "0     AG1000G-AO            81     3.0\n",
              "1   AG1000G-BF-A           181     3.0\n",
              "2   AG1000G-BF-B           102     3.0\n",
              "3   AG1000G-BF-C            13     3.0\n",
              "4     AG1000G-CD            76     3.0\n",
              "5     AG1000G-CF            73     3.0\n",
              "6     AG1000G-CI            80     3.0\n",
              "7   AG1000G-CM-A           303     3.0\n",
              "8   AG1000G-CM-B            97     3.0\n",
              "9   AG1000G-CM-C            44     3.0\n",
              "10    AG1000G-FR            23     3.0\n",
              "11  AG1000G-GA-A            69     3.0\n",
              "12    AG1000G-GH           100     3.0\n",
              "13  AG1000G-GM-A            74     3.0\n",
              "14  AG1000G-GM-B            31     3.0\n",
              "15  AG1000G-GM-C           174     3.0\n",
              "16  AG1000G-GN-A            45     3.0\n",
              "17  AG1000G-GN-B           185     3.0\n",
              "18    AG1000G-GQ            10     3.0\n",
              "19    AG1000G-GW           101     3.0\n",
              "20    AG1000G-KE            86     3.0\n",
              "21  AG1000G-ML-A            60     3.0\n",
              "22  AG1000G-ML-B            71     3.0\n",
              "23    AG1000G-MW            41     3.0\n",
              "24    AG1000G-MZ            74     3.0\n",
              "25    AG1000G-TZ           300     3.0\n",
              "26    AG1000G-UG           290     3.0\n",
              "27     AG1000G-X           297     3.0"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ],
      "source": [
        "ag3.sample_sets(release=\"3.0\")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "JgzBJZ35uN_m",
      "metadata": {
        "id": "JgzBJZ35uN_m"
      },
      "source": [
        "To keep things simple, for this module we'll focus on mosquitoes from Burkina Faso. There are three sample sets in the Ag3.0 resource providing data on mosquitoes from Burkina Faso."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "id": "I4XmC_CIuX0I",
      "metadata": {
        "id": "I4XmC_CIuX0I"
      },
      "outputs": [],
      "source": [
        "sample_sets = [\"AG1000G-BF-A\", \"AG1000G-BF-B\", \"AG1000G-BF-C\"]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3B0z9b0XuKoo",
      "metadata": {
        "id": "3B0z9b0XuKoo"
      },
      "source": [
        "OK, now we're ready to run the analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "id": "KfjWX8BAtj7h",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 536
        },
        "id": "KfjWX8BAtj7h",
        "outputId": "90882a5b-ba24-43e3-9e87-134645bc80ea"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-1ea0aa58-0390-4c26-9dcd-b18882b09019\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>pass_gamb_colu_arab</th>\n",
              "      <th>pass_gamb_colu</th>\n",
              "      <th>pass_arab</th>\n",
              "      <th>frq_BF-09_gamb_2012</th>\n",
              "      <th>frq_BF-09_colu_2012</th>\n",
              "      <th>frq_BF-09_colu_2014</th>\n",
              "      <th>frq_BF-09_gamb_2014</th>\n",
              "      <th>frq_BF-07_gamb_2004</th>\n",
              "      <th>max_af</th>\n",
              "      <th>effect</th>\n",
              "      <th>impact</th>\n",
              "      <th>ref_codon</th>\n",
              "      <th>alt_codon</th>\n",
              "      <th>aa_pos</th>\n",
              "      <th>ref_aa</th>\n",
              "      <th>alt_aa</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>contig</th>\n",
              "      <th>position</th>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"11\" valign=\"top\">2L</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">2358158</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>M1L</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>START_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>Atg</td>\n",
              "      <td>Ctg</td>\n",
              "      <td>1.0</td>\n",
              "      <td>M</td>\n",
              "      <td>L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>M1L</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>START_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>Atg</td>\n",
              "      <td>Ttg</td>\n",
              "      <td>1.0</td>\n",
              "      <td>M</td>\n",
              "      <td>L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>M1V</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>START_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>Atg</td>\n",
              "      <td>Gtg</td>\n",
              "      <td>1.0</td>\n",
              "      <td>M</td>\n",
              "      <td>V</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2358159</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">T</th>\n",
              "      <th>A</th>\n",
              "      <th>M1K</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>aTg</td>\n",
              "      <td>aAg</td>\n",
              "      <td>1.0</td>\n",
              "      <td>M</td>\n",
              "      <td>K</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C</th>\n",
              "      <th>M1T</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>aTg</td>\n",
              "      <td>aCg</td>\n",
              "      <td>1.0</td>\n",
              "      <td>M</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2431616</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">G</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119S</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>tGa</td>\n",
              "      <td>tCa</td>\n",
              "      <td>2119.0</td>\n",
              "      <td>*</td>\n",
              "      <td>S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119L</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>tGa</td>\n",
              "      <td>tTa</td>\n",
              "      <td>2119.0</td>\n",
              "      <td>*</td>\n",
              "      <td>L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">2431617</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119C</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>tgA</td>\n",
              "      <td>tgC</td>\n",
              "      <td>2119.0</td>\n",
              "      <td>*</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119C</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>tgA</td>\n",
              "      <td>tgT</td>\n",
              "      <td>2119.0</td>\n",
              "      <td>*</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>*2119W</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "      <td>tgA</td>\n",
              "      <td>tgG</td>\n",
              "      <td>2119.0</td>\n",
              "      <td>*</td>\n",
              "      <td>W</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>209001 rows √ó 16 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-1ea0aa58-0390-4c26-9dcd-b18882b09019')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-1ea0aa58-0390-4c26-9dcd-b18882b09019 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-1ea0aa58-0390-4c26-9dcd-b18882b09019');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                                 pass_gamb_colu_arab  ...  alt_aa\n",
              "contig position ref_allele alt_allele aa_change                       ...        \n",
              "2L     2358158  A          C          M1L                       True  ...       L\n",
              "                           T          M1L                       True  ...       L\n",
              "                           G          M1V                       True  ...       V\n",
              "       2358159  T          A          M1K                       True  ...       K\n",
              "                           C          M1T                       True  ...       T\n",
              "...                                                              ...  ...     ...\n",
              "       2431616  G          C          *2119S                    True  ...       S\n",
              "                           T          *2119L                    True  ...       L\n",
              "       2431617  A          C          *2119C                    True  ...       C\n",
              "                           T          *2119C                    True  ...       C\n",
              "                           G          *2119W                    True  ...       W\n",
              "\n",
              "[209001 rows x 16 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ],
      "source": [
        "snp_allele_freqs_df = ag3.snp_allele_frequencies(\n",
        "    transcript=transcript, \n",
        "    cohorts=cohorts, \n",
        "    sample_sets=sample_sets, \n",
        "    drop_invariant=False,\n",
        ")\n",
        "snp_allele_freqs_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "YpX2M5sbu24x",
      "metadata": {
        "id": "YpX2M5sbu24x"
      },
      "source": [
        "The output from this function is a pandas DataFrame, where each row provides information about a SNP in the *Vgsc* gene.\n",
        "\n",
        "Before we go further, to improve our understanding of what these data, let's look at some background."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "PBco7cz5viIw",
      "metadata": {
        "id": "PBco7cz5viIw"
      },
      "source": [
        "## Grouping samples into cohorts\n",
        "\n",
        "The MalariaGEN Ag3.0 data resource contain mosquito samples collected across large spatial and temporal scales, and from different mosquito species. When we want to run population genetic analyses on datasets like these, the data must be divided into biologically relevant **cohorts**, where a cohort is simply a group of samples we want to analyse together. \n",
        "\n",
        "To help define cohorts and analyse these data, we have added some metadata for each sample about its time and place of collection and its species: \n",
        "\n",
        "- **Spatially** - For most analyses we use administrative divisions to group the samples into cohorts. These give two levels of spatial resolution, where admin level 1 divides each country into a few large regions, while admin level 2 provides finer scale divisions such as provinces.\n",
        "\n",
        "- **Temporally** - For each sample we provide the year and month of collection. Depending on your analysis, you can choose to group samples by year or by year and by month, although note that for some samples the collection month is missing.\n",
        "\n",
        "- **Taxonomically** - Ag3.0 contains samples from different species in the *Anopheles gambiae* complex. To help with grouping by taxon, we have included a \"taxon\" field in the sample metadata.\n",
        "   \n",
        "Using these three dimensions, we have pre-defined four **cohort sets**, each of which groups samples into cohorts at different levels of spatio-temporal resolution. Within all cohort sets, samples are further subdivided by taxon.\n",
        "\n",
        "- **admin1_year** - Cohorts obtained by grouping samples by admin level 1, collection year and taxon.\n",
        "- **admin1_month** - Cohorts obtained by grouping samples by admin level 1, collection year and month, and taxon.\n",
        "- **admin2_year** - Cohorts obtained by grouping samples by admin level 2, collection year and taxon.\n",
        "- **admin2_month** - Cohorts obtained by grouping samples by admin level 2, collection year and month, and taxon.\n",
        "\n",
        "Remember above we chose to use the \"admin1_year\" cohorts for our *Vgsc* analysis. "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "id": "h0xhDXtq2rEv",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "h0xhDXtq2rEv",
        "outputId": "367e56e1-5413-4409-a7a8-d43ffe3661c8"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'admin1_year'"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ],
      "source": [
        "cohorts"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "BNT804jI2rfc",
      "metadata": {
        "id": "BNT804jI2rfc"
      },
      "source": [
        "Let's now use this to understand the frequency columns in the SNP allele frequencies DataFrame. Here are the frequency column names:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "id": "uzuWWMx222Zd",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uzuWWMx222Zd",
        "outputId": "06502ea2-7ac0-478b-8ff6-ef5fa7b9a827"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['frq_BF-09_gamb_2012',\n",
              " 'frq_BF-09_colu_2012',\n",
              " 'frq_BF-09_colu_2014',\n",
              " 'frq_BF-09_gamb_2014',\n",
              " 'frq_BF-07_gamb_2004']"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ],
      "source": [
        "frequency_columns = [\n",
        "    col for col in snp_allele_freqs_df.columns \n",
        "    if col.startswith(\"frq_\")\n",
        "]\n",
        "frequency_columns"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "JklxNfiK2-LV",
      "metadata": {
        "id": "JklxNfiK2-LV"
      },
      "source": [
        "Here \"frq_\" is used to mean that these columns contain frequencies.\n",
        "\n",
        "The second part of the column name is either \"BF-09\" or \"BF-07\". These are [standard identifiers](https://en.wikipedia.org/wiki/ISO_3166-2) that refer to level 1 administrative divisions within countries, often called regions. \"BF-07\" is the Centre-Sud region within Burkina Faso, and \"BF-09\" is the Haut-Bassins region.\n",
        "\n",
        "The third part refers to the species, and in this example is either \"gamb\" meaning *Anopheles gambiae* or \"colu\" meaning *Anopheles coluzzii*.\n",
        "\n",
        "The final part is the year of collection, which in this example ie either 2004, 2012 or 2014.\n",
        "\n",
        "Thus, e.g., \"frq_BF-09_gamb_2012\" means the frequency of the allele within the cohort of *Anopheles gambiae* mosquitoes from the BF-09 (Haut-Bassins) region collected in 2012."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "AbBbuyxxDF63",
      "metadata": {
        "id": "AbBbuyxxDF63"
      },
      "source": [
        "## Single nucleotide polymorphisms (SNPs)\n",
        "\n",
        "In the last module we learnt about reference genomes and gene annotations. Now one of the cool things we can do once we have an _Anopheles gambiae_ reference genome, is sequence the genomes of wild mosquitoes and compare them to the reference genome. By aligning an individual's sequencing reads to the reference, we can call its **genotype**. \n",
        "\n",
        "The genotype is derived from how similar the individual is to the reference genome at each **nucleotide position**. \n",
        "\n",
        "Where a nucleotide matches the reference genome, it will be called a **reference allele**, where \"allele\" is just a term for any kind of genetic variant. Where a nucleotide differs from the reference genome, it will be called an **alternative allele**. \n",
        "\n",
        "Since, there are four possible nucleotides and one of these is the reference, there are always three possible alternative alleles for any given site. Another term for a genetic difference where two individuals differ at a single nucleotide position is **single nucleotide polymorphism** or **SNP**."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "id": "18zlgylaDtfL",
      "metadata": {
        "id": "18zlgylaDtfL",
        "tags": [
          "remove-input"
        ],
        "outputId": "2c364897-cdad-4551-ea10-5986c6788089",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 721
        }
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<iframe frameborder=\"0\" style=\"width:100%;height:700px;\" src=\"https://viewer.diagrams.net/?tags=%7B%7D&highlight=000000&edit=_blank&layers=1&nav=1#G1MXxUT74w6lQ7FRppVmza4M4cdS_ZecNw\"></iframe>"
            ],
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "%%html\n",
        "<iframe frameborder=\"0\" style=\"width:100%;height:700px;\" src=\"https://viewer.diagrams.net/?tags=%7B%7D&highlight=000000&edit=_blank&layers=1&nav=1#G1MXxUT74w6lQ7FRppVmza4M4cdS_ZecNw\"></iframe>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "AYLAnQ7VEmbP",
      "metadata": {
        "id": "AYLAnQ7VEmbP"
      },
      "source": [
        "This SNP information is displayed in the first four columns of the `snp_allele_frequencies()` index in the output dataframe. We can see each alternative allele represented by it's own row in our output."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "id": "IH3EDr_NEpLd",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 437
        },
        "id": "IH3EDr_NEpLd",
        "outputId": "1a8e6152-bd66-4fcb-f272-2b3d3a1f006f"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-bdd1c5e4-f938-4969-a2f5-fa292f83631e\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>contig</th>\n",
              "      <th>position</th>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"11\" valign=\"top\">2L</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">2358158</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>M1L</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>M1L</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>M1V</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2358159</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">T</th>\n",
              "      <th>A</th>\n",
              "      <th>M1K</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C</th>\n",
              "      <th>M1T</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2431616</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">G</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119S</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119L</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">2431617</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119C</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119C</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>*2119W</th>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>209001 rows √ó 0 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-bdd1c5e4-f938-4969-a2f5-fa292f83631e')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-bdd1c5e4-f938-4969-a2f5-fa292f83631e button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-bdd1c5e4-f938-4969-a2f5-fa292f83631e');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: []\n",
              "Index: [(2L, 2358158, A, C, M1L), (2L, 2358158, A, T, M1L), (2L, 2358158, A, G, M1V), (2L, 2358159, T, A, M1K), (2L, 2358159, T, C, M1T), (2L, 2358159, T, G, M1R), (2L, 2358160, G, A, M1I), (2L, 2358160, G, C, M1I), (2L, 2358160, G, T, M1I), (2L, 2358161, A, C, T2P), (2L, 2358161, A, T, T2S), (2L, 2358161, A, G, T2A), (2L, 2358162, C, A, T2N), (2L, 2358162, C, T, T2I), (2L, 2358162, C, G, T2S), (2L, 2358163, C, A, T2T), (2L, 2358163, C, T, T2T), (2L, 2358163, C, G, T2T), (2L, 2358164, G, A, E3K), (2L, 2358164, G, C, E3Q), (2L, 2358164, G, T, E3*), (2L, 2358165, A, C, E3A), (2L, 2358165, A, T, E3V), (2L, 2358165, A, G, E3G), (2L, 2358166, A, C, E3D), (2L, 2358166, A, T, E3D), (2L, 2358166, A, G, E3E), (2L, 2358167, G, A, D4N), (2L, 2358167, G, C, D4H), (2L, 2358167, G, T, D4Y), (2L, 2358168, A, C, D4A), (2L, 2358168, A, T, D4V), (2L, 2358168, A, G, D4G), (2L, 2358169, C, A, D4E), (2L, 2358169, C, T, D4D), (2L, 2358169, C, G, D4E), (2L, 2358170, T, A, S5T), (2L, 2358170, T, C, S5P), (2L, 2358170, T, G, S5A), (2L, 2358171, C, A, S5Y), (2L, 2358171, C, T, S5F), (2L, 2358171, C, G, S5C), (2L, 2358172, C, A, S5S), (2L, 2358172, C, T, S5S), (2L, 2358172, C, G, S5S), (2L, 2358173, G, A, D6N), (2L, 2358173, G, C, D6H), (2L, 2358173, G, T, D6Y), (2L, 2358174, A, C, D6A), (2L, 2358174, A, T, D6V), (2L, 2358174, A, G, D6G), (2L, 2358175, T, A, D6E), (2L, 2358175, T, C, D6D), (2L, 2358175, T, G, D6E), (2L, 2358176, T, A, S7T), (2L, 2358176, T, C, S7P), (2L, 2358176, T, G, S7A), (2L, 2358177, C, A, S7*), (2L, 2358177, C, T, S7L), (2L, 2358177, C, G, S7W), (2L, 2358178, G, A, S7S), (2L, 2358178, G, C, S7S), (2L, 2358178, G, T, S7S), (2L, 2358179, A, C, I8L), (2L, 2358179, A, T, I8L), (2L, 2358179, A, G, I8V), (2L, 2358180, T, A, I8K), (2L, 2358180, T, C, I8T), (2L, 2358180, T, G, I8R), (2L, 2358181, A, C, I8I), (2L, 2358181, A, T, I8I), (2L, 2358181, A, G, I8M), (2L, 2358182, T, A, S9T), (2L, 2358182, T, C, S9P), (2L, 2358182, T, G, S9A), (2L, 2358183, C, A, S9Y), (2L, 2358183, C, T, S9F), (2L, 2358183, C, G, S9C), (2L, 2358184, T, A, S9S), (2L, 2358184, T, C, S9S), (2L, 2358184, T, G, S9S), (2L, 2358185, G, A, E10K), (2L, 2358185, G, C, E10Q), (2L, 2358185, G, T, E10*), (2L, 2358186, A, C, E10A), (2L, 2358186, A, T, E10V), (2L, 2358186, A, G, E10G), (2L, 2358187, G, A, E10E), (2L, 2358187, G, C, E10D), (2L, 2358187, G, T, E10D), (2L, 2358188, G, A, E11K), (2L, 2358188, G, C, E11Q), (2L, 2358188, G, T, E11*), (2L, 2358189, A, C, E11A), (2L, 2358189, A, T, E11V), (2L, 2358189, A, G, E11G), (2L, 2358190, A, C, E11D), (2L, 2358190, A, T, E11D), (2L, 2358190, A, G, E11E), (2L, 2358191, G, A, E12K), ...]\n",
              "\n",
              "[209001 rows x 0 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ],
      "source": [
        "snp_allele_freqs_df[[]]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "L46DP8FyFwGP",
      "metadata": {
        "id": "L46DP8FyFwGP"
      },
      "source": [
        "## SNP effects - some SNPs may be more interesting than others\n",
        "\n",
        "SNPs can have different effects depending on what the nucleotide change is and where in the genome it occurs.\n",
        "\n",
        "In this analysis, we are interested in SNPs that affect protein structure, specifically those which will change the voltage-gated sodium-channel and could affect protein function and therefore insecticide resistance phenotype, so we need to look within the coding sequences (CDS) of the _Vgsc_ gene.\n",
        "\n",
        "However, not all SNPs which fall in CDSs cause protein changes, because the genetic code has some redundancy, meaning that different nucleotide sequences can encode the same amino acid. If a SNP in a CDS does change the amino acid, it is called a **non-synonymous** (NS) or **missense** SNP, and if the SNP does not change the amino acid, it is called a **synonymous** SNP.  \n",
        "\n",
        "Manually predicting SNP effects is quite involved, but `snp_allele_frequencies()` can predict them for us, let's look at how this is represented in our output DataFrame.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "id": "UeWw_boUGkKa",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 455
        },
        "id": "UeWw_boUGkKa",
        "outputId": "1c238933-901c-469c-fe16-ac6b7443710e"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-17a04152-4465-42c7-a50e-1cab5ad9e17d\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>effect</th>\n",
              "      <th>impact</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>contig</th>\n",
              "      <th>position</th>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"11\" valign=\"top\">2L</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">2358158</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>M1L</th>\n",
              "      <td>START_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>M1L</th>\n",
              "      <td>START_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>M1V</th>\n",
              "      <td>START_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2358159</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">T</th>\n",
              "      <th>A</th>\n",
              "      <th>M1K</th>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C</th>\n",
              "      <th>M1T</th>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2431616</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">G</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119S</th>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119L</th>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">2431617</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119C</th>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119C</th>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>*2119W</th>\n",
              "      <td>STOP_LOST</td>\n",
              "      <td>HIGH</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>209001 rows √ó 2 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-17a04152-4465-42c7-a50e-1cab5ad9e17d')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-17a04152-4465-42c7-a50e-1cab5ad9e17d button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-17a04152-4465-42c7-a50e-1cab5ad9e17d');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                                                effect    impact\n",
              "contig position ref_allele alt_allele aa_change                                 \n",
              "2L     2358158  A          C          M1L                   START_LOST      HIGH\n",
              "                           T          M1L                   START_LOST      HIGH\n",
              "                           G          M1V                   START_LOST      HIGH\n",
              "       2358159  T          A          M1K        NON_SYNONYMOUS_CODING  MODERATE\n",
              "                           C          M1T        NON_SYNONYMOUS_CODING  MODERATE\n",
              "...                                                                ...       ...\n",
              "       2431616  G          C          *2119S                 STOP_LOST      HIGH\n",
              "                           T          *2119L                 STOP_LOST      HIGH\n",
              "       2431617  A          C          *2119C                 STOP_LOST      HIGH\n",
              "                           T          *2119C                 STOP_LOST      HIGH\n",
              "                           G          *2119W                 STOP_LOST      HIGH\n",
              "\n",
              "[209001 rows x 2 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ],
      "source": [
        "snp_effects_df = snp_allele_freqs_df[[\"effect\", \"impact\"]]\n",
        "snp_effects_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "dX4o4gPSG0rQ",
      "metadata": {
        "id": "dX4o4gPSG0rQ"
      },
      "source": [
        "Let's look specifically at the genomic position where a SNP occurs which causes an insecticide-resistance mutation, also known as \"kdr\". "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "id": "LTmD08AfGNYP",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 175
        },
        "id": "LTmD08AfGNYP",
        "outputId": "303163fc-38d4-4bb0-a564-3a271d40c480"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-4ccd2bc2-96eb-4959-b973-c62f9924f134\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>effect</th>\n",
              "      <th>impact</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>L995F</th>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>L995F</th>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>L995L</th>\n",
              "      <td>SYNONYMOUS_CODING</td>\n",
              "      <td>LOW</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4ccd2bc2-96eb-4959-b973-c62f9924f134')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-4ccd2bc2-96eb-4959-b973-c62f9924f134 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-4ccd2bc2-96eb-4959-b973-c62f9924f134');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                                effect    impact\n",
              "ref_allele alt_allele aa_change                                 \n",
              "A          C          L995F      NON_SYNONYMOUS_CODING  MODERATE\n",
              "           T          L995F      NON_SYNONYMOUS_CODING  MODERATE\n",
              "           G          L995L          SYNONYMOUS_CODING       LOW"
            ]
          },
          "metadata": {},
          "execution_count": 16
        }
      ],
      "source": [
        "snp_effects_df.loc[(\"2L\", 2_422_652)]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "kaOYrxB-G_fO",
      "metadata": {
        "id": "kaOYrxB-G_fO"
      },
      "source": [
        "For interest, let's count the number of SNPs we have by their effect."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "id": "dy413q3jHF1o",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dy413q3jHF1o",
        "outputId": "53d69165-a001-476c-f74a-7abaedd7afb9"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "effect                 impact  \n",
              "INTRONIC               MODIFIER    9451\n",
              "NON_SYNONYMOUS_CODING  MODERATE     121\n",
              "SPLICE_CORE            HIGH           8\n",
              "SPLICE_REGION          MODERATE       9\n",
              "STOP_GAINED            HIGH          14\n",
              "SYNONYMOUS_CODING      LOW           56\n",
              "dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 17
        }
      ],
      "source": [
        "snp_allele_freqs_df.query(\"max_af > 0\").groupby([\"effect\", \"impact\"]).size()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "PPCxaa3xHEJg",
      "metadata": {
        "id": "PPCxaa3xHEJg"
      },
      "source": [
        "## SNP allele frequencies\n",
        "\n",
        "Identifying the presence or absence of SNPs in wild caught mosquitoes is interesting, but the real value in generating SNP genotypes from large spatiotemporal collections of mosquitoes comes from the ability to see how groups of samples (cohorts) differ between geographical locations, species, and over time.  \n",
        "\n",
        "One way to compare SNP differences between cohorts is to calculate and compare **SNP allele frequencies** at each position in the genome by dividing the number of times each SNP allele is found in the cohort by the total number of individuals present in the cohort (multiplied by 2 because each individual mosquito is diploid and so carries two genome copies)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "id": "9NMPBIn2Hitt",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 721
        },
        "id": "9NMPBIn2Hitt",
        "outputId": "39cafb8e-d885-4387-cb9e-a5fc2861cd64"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<iframe frameborder=\"0\" style=\"width:100%;height:700px;\" src=\"https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=allele_freqs.drawio#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D1vbRdL36K4NAeQWOF59o4KR723rZrBnak%26export%3Ddownload\"></iframe>"
            ],
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "%%html\n",
        "<iframe frameborder=\"0\" style=\"width:100%;height:700px;\" src=\"https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=allele_freqs.drawio#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D1vbRdL36K4NAeQWOF59o4KR723rZrBnak%26export%3Ddownload\"></iframe>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ARwbodAAHtec",
      "metadata": {
        "id": "ARwbodAAHtec"
      },
      "source": [
        "Let's take another look at the allele frequencies we computed above, focusing just on the frequency columns."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "id": "hsx08pBVH1Gu",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 516
        },
        "id": "hsx08pBVH1Gu",
        "outputId": "eded9ee0-5821-4dcc-ff1e-1bcf828b3b8b"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-82aee72d-8e44-4381-af62-2337d00f31bd\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>frq_BF-09_gamb_2012</th>\n",
              "      <th>frq_BF-09_colu_2012</th>\n",
              "      <th>frq_BF-09_colu_2014</th>\n",
              "      <th>frq_BF-09_gamb_2014</th>\n",
              "      <th>frq_BF-07_gamb_2004</th>\n",
              "      <th>max_af</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>contig</th>\n",
              "      <th>position</th>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"11\" valign=\"top\">2L</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">2358158</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>M1L</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>M1L</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>M1V</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2358159</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">T</th>\n",
              "      <th>A</th>\n",
              "      <th>M1K</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C</th>\n",
              "      <th>M1T</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">2431616</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">G</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119S</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119L</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">2431617</th>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>*2119C</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>*2119C</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>*2119W</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>209001 rows √ó 6 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-82aee72d-8e44-4381-af62-2337d00f31bd')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-82aee72d-8e44-4381-af62-2337d00f31bd button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-82aee72d-8e44-4381-af62-2337d00f31bd');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                                 frq_BF-09_gamb_2012  ...  max_af\n",
              "contig position ref_allele alt_allele aa_change                       ...        \n",
              "2L     2358158  A          C          M1L                        0.0  ...     0.0\n",
              "                           T          M1L                        0.0  ...     0.0\n",
              "                           G          M1V                        0.0  ...     0.0\n",
              "       2358159  T          A          M1K                        0.0  ...     0.0\n",
              "                           C          M1T                        0.0  ...     0.0\n",
              "...                                                              ...  ...     ...\n",
              "       2431616  G          C          *2119S                     0.0  ...     0.0\n",
              "                           T          *2119L                     0.0  ...     0.0\n",
              "       2431617  A          C          *2119C                     0.0  ...     0.0\n",
              "                           T          *2119C                     0.0  ...     0.0\n",
              "                           G          *2119W                     0.0  ...     0.0\n",
              "\n",
              "[209001 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 19
        }
      ],
      "source": [
        "snp_allele_freqs_df[frequency_columns + ['max_af']]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fvqO9se6IDwO",
      "metadata": {
        "id": "fvqO9se6IDwO"
      },
      "source": [
        "And let's inspect the frequencies for SNPs at a specific genomic position of interest."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "id": "wFB2RJJwIHjJ",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 236
        },
        "id": "wFB2RJJwIHjJ",
        "outputId": "a931816d-c67d-4ff7-cbbe-62b4099bba04"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>frq_BF-09_gamb_2012</th>\n",
              "      <th>frq_BF-09_colu_2012</th>\n",
              "      <th>frq_BF-09_colu_2014</th>\n",
              "      <th>frq_BF-09_gamb_2014</th>\n",
              "      <th>frq_BF-07_gamb_2004</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"3\" valign=\"top\">A</th>\n",
              "      <th>C</th>\n",
              "      <th>L995F</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>L995F</th>\n",
              "      <td>1.0</td>\n",
              "      <td>0.865854</td>\n",
              "      <td>0.886792</td>\n",
              "      <td>1.0</td>\n",
              "      <td>0.076923</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G</th>\n",
              "      <th>L995L</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                 frq_BF-09_gamb_2012  ...  frq_BF-07_gamb_2004\n",
              "ref_allele alt_allele aa_change                       ...                     \n",
              "A          C          L995F                      0.0  ...             0.000000\n",
              "           T          L995F                      1.0  ...             0.076923\n",
              "           G          L995L                      0.0  ...             0.000000\n",
              "\n",
              "[3 rows x 5 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ],
      "source": [
        "snp_allele_freqs_df.loc[(\"2L\", 2_422_652), frequency_columns]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ST9mnV9NISo9",
      "metadata": {
        "id": "ST9mnV9NISo9"
      },
      "source": [
        "## Visualising SNP allele frequencies\n",
        "\n",
        "To make our SNP allele frequencies DataFrame easier to interpret, we can filter it down to just non-synonymous SNPs that are at frequency above 5% in at least one of our cohorts."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "id": "EwWK09z_Ibt2",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 570
        },
        "id": "EwWK09z_Ibt2",
        "outputId": "d08a3a4e-12f8-4397-b805-b5a1749e403e"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>pass_gamb_colu_arab</th>\n",
              "      <th>pass_gamb_colu</th>\n",
              "      <th>pass_arab</th>\n",
              "      <th>frq_BF-09_gamb_2012</th>\n",
              "      <th>frq_BF-09_colu_2012</th>\n",
              "      <th>frq_BF-09_colu_2014</th>\n",
              "      <th>frq_BF-09_gamb_2014</th>\n",
              "      <th>frq_BF-07_gamb_2004</th>\n",
              "      <th>max_af</th>\n",
              "      <th>effect</th>\n",
              "      <th>impact</th>\n",
              "      <th>ref_codon</th>\n",
              "      <th>alt_codon</th>\n",
              "      <th>aa_pos</th>\n",
              "      <th>ref_aa</th>\n",
              "      <th>alt_aa</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>contig</th>\n",
              "      <th>position</th>\n",
              "      <th>ref_allele</th>\n",
              "      <th>alt_allele</th>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"13\" valign=\"top\">2L</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">2391228</th>\n",
              "      <th rowspan=\"2\" valign=\"top\">G</th>\n",
              "      <th>C</th>\n",
              "      <th>V402L</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.067073</td>\n",
              "      <td>0.028302</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.067073</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>Gta</td>\n",
              "      <td>Cta</td>\n",
              "      <td>402.0</td>\n",
              "      <td>V</td>\n",
              "      <td>L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T</th>\n",
              "      <th>V402L</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.054878</td>\n",
              "      <td>0.084906</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.084906</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>Gta</td>\n",
              "      <td>Tta</td>\n",
              "      <td>402.0</td>\n",
              "      <td>V</td>\n",
              "      <td>L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2416980</th>\n",
              "      <th>C</th>\n",
              "      <th>T</th>\n",
              "      <th>T791M</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.163265</td>\n",
              "      <td>0.018293</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>aCg</td>\n",
              "      <td>aTg</td>\n",
              "      <td>791.0</td>\n",
              "      <td>T</td>\n",
              "      <td>M</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2422652</th>\n",
              "      <th>A</th>\n",
              "      <th>T</th>\n",
              "      <th>L995F</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.865854</td>\n",
              "      <td>0.886792</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.076923</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>ttA</td>\n",
              "      <td>ttT</td>\n",
              "      <td>995.0</td>\n",
              "      <td>L</td>\n",
              "      <td>F</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2429617</th>\n",
              "      <th>T</th>\n",
              "      <th>C</th>\n",
              "      <th>I1527T</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "      <td>0.113208</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>aTt</td>\n",
              "      <td>aCt</td>\n",
              "      <td>1527.0</td>\n",
              "      <td>I</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2429745</th>\n",
              "      <th>A</th>\n",
              "      <th>T</th>\n",
              "      <th>N1570Y</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.209184</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>0.320755</td>\n",
              "      <td>0.141304</td>\n",
              "      <td>0.038462</td>\n",
              "      <td>0.320755</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>Aat</td>\n",
              "      <td>Tat</td>\n",
              "      <td>1570.0</td>\n",
              "      <td>N</td>\n",
              "      <td>Y</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2429897</th>\n",
              "      <th>A</th>\n",
              "      <th>G</th>\n",
              "      <th>E1597G</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.066327</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.032609</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.066327</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>gAa</td>\n",
              "      <td>gGa</td>\n",
              "      <td>1597.0</td>\n",
              "      <td>E</td>\n",
              "      <td>G</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2429915</th>\n",
              "      <th>A</th>\n",
              "      <th>C</th>\n",
              "      <th>K1603T</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.054878</td>\n",
              "      <td>0.056604</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.056604</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>aAg</td>\n",
              "      <td>aCg</td>\n",
              "      <td>1603.0</td>\n",
              "      <td>K</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2430424</th>\n",
              "      <th>G</th>\n",
              "      <th>T</th>\n",
              "      <th>A1746S</th>\n",
              "      <td>False</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "      <td>0.153061</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>Gcc</td>\n",
              "      <td>Tcc</td>\n",
              "      <td>1746.0</td>\n",
              "      <td>A</td>\n",
              "      <td>S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2430863</th>\n",
              "      <th>T</th>\n",
              "      <th>C</th>\n",
              "      <th>I1868T</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.206522</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>aTa</td>\n",
              "      <td>aCa</td>\n",
              "      <td>1868.0</td>\n",
              "      <td>I</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2430880</th>\n",
              "      <th>C</th>\n",
              "      <th>T</th>\n",
              "      <th>P1874S</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.213415</td>\n",
              "      <td>0.169811</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.213415</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>Cct</td>\n",
              "      <td>Tct</td>\n",
              "      <td>1874.0</td>\n",
              "      <td>P</td>\n",
              "      <td>S</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2430881</th>\n",
              "      <th>C</th>\n",
              "      <th>T</th>\n",
              "      <th>P1874L</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.224490</td>\n",
              "      <td>0.073171</td>\n",
              "      <td>0.056604</td>\n",
              "      <td>0.260870</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.260870</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>cCt</td>\n",
              "      <td>cTt</td>\n",
              "      <td>1874.0</td>\n",
              "      <td>P</td>\n",
              "      <td>L</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2431061</th>\n",
              "      <th>C</th>\n",
              "      <th>T</th>\n",
              "      <th>A1934V</th>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "      <td>0.132075</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.132075</td>\n",
              "      <td>NON_SYNONYMOUS_CODING</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>gCt</td>\n",
              "      <td>gTt</td>\n",
              "      <td>1934.0</td>\n",
              "      <td>A</td>\n",
              "      <td>V</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "                                                 pass_gamb_colu_arab  ...  alt_aa\n",
              "contig position ref_allele alt_allele aa_change                       ...        \n",
              "2L     2391228  G          C          V402L                     True  ...       L\n",
              "                           T          V402L                     True  ...       L\n",
              "       2416980  C          T          T791M                     True  ...       M\n",
              "       2422652  A          T          L995F                     True  ...       F\n",
              "       2429617  T          C          I1527T                    True  ...       T\n",
              "       2429745  A          T          N1570Y                    True  ...       Y\n",
              "       2429897  A          G          E1597G                    True  ...       G\n",
              "       2429915  A          C          K1603T                    True  ...       T\n",
              "       2430424  G          T          A1746S                   False  ...       S\n",
              "       2430863  T          C          I1868T                    True  ...       T\n",
              "       2430880  C          T          P1874S                    True  ...       S\n",
              "       2430881  C          T          P1874L                    True  ...       L\n",
              "       2431061  C          T          A1934V                    True  ...       V\n",
              "\n",
              "[13 rows x 16 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 21
        }
      ],
      "source": [
        "ns_snps_df = snp_allele_freqs_df.query(\"effect == 'NON_SYNONYMOUS_CODING' and max_af >= 0.05\")\n",
        "ns_snps_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ARTAdQvoIgL8",
      "metadata": {
        "id": "ARTAdQvoIgL8"
      },
      "source": [
        "To make things even clearer, we have included a heatmap plotting function to style our filtered DataFrame, called [`plot_frequencies_heatmap()`](https://malariagen.github.io/vector-data/ag3/api.html#plot-frequencies-heatmap)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "id": "krxkMyJZIpUh",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 542
        },
        "id": "krxkMyJZIpUh",
        "outputId": "3eda942f-ba85-476d-8565-1bdb339a3cb4"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<html>\n",
              "<head><meta charset=\"utf-8\" /></head>\n",
              "<body>\n",
              "    <div>            <script src=\"https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG\"></script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script>                <script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script>\n",
              "        <script src=\"https://cdn.plot.ly/plotly-2.8.3.min.js\"></script>                <div id=\"37d951e7-ec15-4ad6-b5ba-f109a526b881\" class=\"plotly-graph-div\" style=\"height:525px; width:600px;\"></div>            <script type=\"text/javascript\">                                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"37d951e7-ec15-4ad6-b5ba-f109a526b881\")) {                    Plotly.newPlot(                        \"37d951e7-ec15-4ad6-b5ba-f109a526b881\",                        [{\"coloraxis\":\"coloraxis\",\"name\":\"0\",\"texttemplate\":\"%{z:.0%}\",\"x\":[\"BF-09_gamb_2012\",\"BF-09_colu_2012\",\"BF-09_colu_2014\",\"BF-09_gamb_2014\",\"BF-07_gamb_2004\"],\"y\":[\"2L, 2391228, G, C, V402L\",\"2L, 2391228, G, T, V402L\",\"2L, 2416980, C, T, T791M\",\"2L, 2422652, A, T, L995F\",\"2L, 2429617, T, C, I1527T\",\"2L, 2429745, A, T, N1570Y\",\"2L, 2429897, A, G, E1597G\",\"2L, 2429915, A, C, K1603T\",\"2L, 2430424, G, T, A1746S\",\"2L, 2430863, T, C, I1868T\",\"2L, 2430880, C, T, P1874S\",\"2L, 2430881, C, T, P1874L\",\"2L, 2431061, C, T, A1934V\"],\"z\":[[0.0,0.06707317073170732,0.02830188679245283,0.0,0.0],[0.0,0.054878048780487805,0.08490566037735849,0.0,0.0],[0.16326530612244897,0.018292682926829267,0.0,0.2391304347826087,0.0],[1.0,0.8658536585365854,0.8867924528301887,1.0,0.07692307692307693],[0.0,0.12195121951219512,0.11320754716981132,0.0,0.0],[0.20918367346938777,0.25,0.32075471698113206,0.14130434782608695,0.038461538461538464],[0.0663265306122449,0.0,0.0,0.03260869565217391,0.0],[0.0,0.054878048780487805,0.05660377358490566,0.0,0.0],[0.15306122448979592,0.0,0.0,0.2391304347826087,0.0],[0.25,0.0,0.0,0.20652173913043478,0.0],[0.0,0.21341463414634146,0.16981132075471697,0.0,0.0],[0.22448979591836735,0.07317073170731707,0.05660377358490566,0.2608695652173913,0.0],[0.0,0.12195121951219512,0.1320754716981132,0.0,0.0]],\"type\":\"heatmap\",\"xaxis\":\"x\",\"yaxis\":\"y\",\"hovertemplate\":\"x: %{x}<br>y: %{y}<br>color: %{z}<extra></extra>\"}],                        {\"template\":{\"data\":{\"bar\":[{\"error_x\":{\"color\":\"#2a3f5f\"},\"error_y\":{\"color\":\"#2a3f5f\"},\"marker\":{\"line\":{\"color\":\"#E5ECF6\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"bar\"}],\"barpolar\":[{\"marker\":{\"line\":{\"color\":\"#E5ECF6\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"barpolar\"}],\"carpet\":[{\"aaxis\":{\"endlinecolor\":\"#2a3f5f\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"#2a3f5f\"},\"baxis\":{\"endlinecolor\":\"#2a3f5f\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"#2a3f5f\"},\"type\":\"carpet\"}],\"choropleth\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"type\":\"choropleth\"}],\"contour\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"contour\"}],\"contourcarpet\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"type\":\"contourcarpet\"}],\"heatmap\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"heatmap\"}],\"heatmapgl\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"heatmapgl\"}],\"histogram\":[{\"marker\":{\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"histogram\"}],\"histogram2d\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"histogram2d\"}],\"histogram2dcontour\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"histogram2dcontour\"}],\"mesh3d\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"type\":\"mesh3d\"}],\"parcoords\":[{\"line\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"parcoords\"}],\"pie\":[{\"automargin\":true,\"type\":\"pie\"}],\"scatter\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatter\"}],\"scatter3d\":[{\"line\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatter3d\"}],\"scattercarpet\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattercarpet\"}],\"scattergeo\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattergeo\"}],\"scattergl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattergl\"}],\"scattermapbox\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattermapbox\"}],\"scatterpolar\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatterpolar\"}],\"scatterpolargl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatterpolargl\"}],\"scatterternary\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatterternary\"}],\"surface\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"surface\"}],\"table\":[{\"cells\":{\"fill\":{\"color\":\"#EBF0F8\"},\"line\":{\"color\":\"white\"}},\"header\":{\"fill\":{\"color\":\"#C8D4E3\"},\"line\":{\"color\":\"white\"}},\"type\":\"table\"}]},\"layout\":{\"annotationdefaults\":{\"arrowcolor\":\"#2a3f5f\",\"arrowhead\":0,\"arrowwidth\":1},\"autotypenumbers\":\"strict\",\"coloraxis\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"colorscale\":{\"diverging\":[[0,\"#8e0152\"],[0.1,\"#c51b7d\"],[0.2,\"#de77ae\"],[0.3,\"#f1b6da\"],[0.4,\"#fde0ef\"],[0.5,\"#f7f7f7\"],[0.6,\"#e6f5d0\"],[0.7,\"#b8e186\"],[0.8,\"#7fbc41\"],[0.9,\"#4d9221\"],[1,\"#276419\"]],\"sequential\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"sequentialminus\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]]},\"colorway\":[\"#636efa\",\"#EF553B\",\"#00cc96\",\"#ab63fa\",\"#FFA15A\",\"#19d3f3\",\"#FF6692\",\"#B6E880\",\"#FF97FF\",\"#FECB52\"],\"font\":{\"color\":\"#2a3f5f\"},\"geo\":{\"bgcolor\":\"white\",\"lakecolor\":\"white\",\"landcolor\":\"#E5ECF6\",\"showlakes\":true,\"showland\":true,\"subunitcolor\":\"white\"},\"hoverlabel\":{\"align\":\"left\"},\"hovermode\":\"closest\",\"mapbox\":{\"style\":\"light\"},\"paper_bgcolor\":\"white\",\"plot_bgcolor\":\"#E5ECF6\",\"polar\":{\"angularaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"},\"bgcolor\":\"#E5ECF6\",\"radialaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"}},\"scene\":{\"xaxis\":{\"backgroundcolor\":\"#E5ECF6\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"},\"yaxis\":{\"backgroundcolor\":\"#E5ECF6\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"},\"zaxis\":{\"backgroundcolor\":\"#E5ECF6\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"}},\"shapedefaults\":{\"line\":{\"color\":\"#2a3f5f\"}},\"ternary\":{\"aaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"},\"baxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"},\"bgcolor\":\"#E5ECF6\",\"caxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"}},\"title\":{\"x\":0.05},\"xaxis\":{\"automargin\":true,\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\",\"title\":{\"standoff\":15},\"zerolinecolor\":\"white\",\"zerolinewidth\":2},\"yaxis\":{\"automargin\":true,\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\",\"title\":{\"standoff\":15},\"zerolinecolor\":\"white\",\"zerolinewidth\":2}}},\"xaxis\":{\"anchor\":\"y\",\"domain\":[0.0,1.0],\"side\":\"top\",\"tickangle\":-90,\"title\":{\"text\":\"cohorts\"}},\"yaxis\":{\"anchor\":\"x\",\"domain\":[0.0,1.0],\"autorange\":\"reversed\"},\"coloraxis\":{\"colorscale\":[[0.0,\"rgb(255,245,240)\"],[0.125,\"rgb(254,224,210)\"],[0.25,\"rgb(252,187,161)\"],[0.375,\"rgb(252,146,114)\"],[0.5,\"rgb(251,106,74)\"],[0.625,\"rgb(239,59,44)\"],[0.75,\"rgb(203,24,29)\"],[0.875,\"rgb(165,15,21)\"],[1.0,\"rgb(103,0,13)\"]],\"cmin\":0,\"cmax\":1,\"colorbar\":{\"title\":{\"text\":\"frequency\"},\"tickvals\":[0,0.2,0.4,0.6,0.8,1.0],\"ticktext\":[\"0%\",\"20%\",\"40%\",\"60%\",\"80%\",\"100%\"]}},\"margin\":{\"t\":60},\"width\":600},                        {\"responsive\": true}                    ).then(function(){\n",
              "                            \n",
              "var gd = document.getElementById('37d951e7-ec15-4ad6-b5ba-f109a526b881');\n",
              "var x = new MutationObserver(function (mutations, observer) {{\n",
              "        var display = window.getComputedStyle(gd).display;\n",
              "        if (!display || display === 'none') {{\n",
              "            console.log([gd, 'removed!']);\n",
              "            Plotly.purge(gd);\n",
              "            observer.disconnect();\n",
              "        }}\n",
              "}});\n",
              "\n",
              "// Listen for the removal of the full notebook cells\n",
              "var notebookContainer = gd.closest('#notebook-container');\n",
              "if (notebookContainer) {{\n",
              "    x.observe(notebookContainer, {childList: true});\n",
              "}}\n",
              "\n",
              "// Listen for the clearing of the current output cell\n",
              "var outputEl = gd.closest('.output');\n",
              "if (outputEl) {{\n",
              "    x.observe(outputEl, {childList: true});\n",
              "}}\n",
              "\n",
              "                        })                };                            </script>        </div>\n",
              "</body>\n",
              "</html>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "ag3.plot_frequencies_heatmap(ns_snps_df, width=600)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "uduTH2qcIvOJ",
      "metadata": {
        "id": "uduTH2qcIvOJ"
      },
      "source": [
        "## Amino acid substitution frequencies\n",
        "\n",
        "You might have noticed that there are two rows with `V402L` in our previous heatmap plot. This is because in Burkina Faso, we find two different alternative alleles at the same genomic position, both causing the same amino acid substitution (valine to leucine).\n",
        "\n",
        "If we are just interested in amino acid change frequencies, for example, when looking at potential insecticide resistance conferring mutations, we might want to combine the frequencies of the two alleles which cause V402L. In this case, we can use the `aa_allele_frequencies()` function in exactly the same way as we used `snp_allele_frequencies()`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "id": "pc8NiqTKI9kh",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 455
        },
        "id": "pc8NiqTKI9kh",
        "outputId": "227385a5-4ba3-42de-e85b-fafd55eea982"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-2bb8a67f-f954-4972-bdbc-769b1d462936\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>frq_BF-09_gamb_2012</th>\n",
              "      <th>frq_BF-09_colu_2012</th>\n",
              "      <th>frq_BF-09_colu_2014</th>\n",
              "      <th>frq_BF-09_gamb_2014</th>\n",
              "      <th>frq_BF-07_gamb_2004</th>\n",
              "      <th>max_af</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>A32V</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.006098</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.006098</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G54C</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.018293</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.010870</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.018293</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>P55L</th>\n",
              "      <td>0.005102</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.010870</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.010870</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>P59T</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.021739</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.021739</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G73D</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.006098</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.006098</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>A2023G</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.038462</td>\n",
              "      <td>0.038462</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>S2037R</th>\n",
              "      <td>0.005102</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.005102</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>I2053V</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.038462</td>\n",
              "      <td>0.038462</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G2055V</th>\n",
              "      <td>0.005102</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.005102</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>V2118A</th>\n",
              "      <td>0.005102</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.005102</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>134 rows √ó 6 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2bb8a67f-f954-4972-bdbc-769b1d462936')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-2bb8a67f-f954-4972-bdbc-769b1d462936 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-2bb8a67f-f954-4972-bdbc-769b1d462936');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "           frq_BF-09_gamb_2012  ...    max_af\n",
              "aa_change                       ...          \n",
              "A32V                  0.000000  ...  0.006098\n",
              "G54C                  0.000000  ...  0.018293\n",
              "P55L                  0.005102  ...  0.010870\n",
              "P59T                  0.000000  ...  0.021739\n",
              "G73D                  0.000000  ...  0.006098\n",
              "...                        ...  ...       ...\n",
              "A2023G                0.000000  ...  0.038462\n",
              "S2037R                0.005102  ...  0.005102\n",
              "I2053V                0.000000  ...  0.038462\n",
              "G2055V                0.005102  ...  0.005102\n",
              "V2118A                0.005102  ...  0.005102\n",
              "\n",
              "[134 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 23
        }
      ],
      "source": [
        "aa_allele_freqs_df = ag3.aa_allele_frequencies(\n",
        "    transcript=transcript, \n",
        "    cohorts=cohorts, \n",
        "    sample_sets=sample_sets\n",
        ")\n",
        "aa_allele_freqs_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "1DhvnM8qJJq3",
      "metadata": {
        "id": "1DhvnM8qJJq3"
      },
      "source": [
        "Let's filter it again to just amino acid changes greater than 5% in at least one cohort. We don't need to filter for non-synonymous mutations this time as this function has already done that for us.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "id": "s4bZbaJgJDnO",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 457
        },
        "id": "s4bZbaJgJDnO",
        "outputId": "ac9593be-9718-4f37-9154-a4484cf0aa58"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-88a4a758-bb32-4762-b31d-203365b338d0\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>frq_BF-09_gamb_2012</th>\n",
              "      <th>frq_BF-09_colu_2012</th>\n",
              "      <th>frq_BF-09_colu_2014</th>\n",
              "      <th>frq_BF-09_gamb_2014</th>\n",
              "      <th>frq_BF-07_gamb_2004</th>\n",
              "      <th>max_af</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>aa_change</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>V402L</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "      <td>0.113208</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>T791M</th>\n",
              "      <td>0.163265</td>\n",
              "      <td>0.018293</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>L995F</th>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.865854</td>\n",
              "      <td>0.886792</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.076923</td>\n",
              "      <td>1.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>I1527T</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "      <td>0.113208</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>N1570Y</th>\n",
              "      <td>0.209184</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>0.320755</td>\n",
              "      <td>0.141304</td>\n",
              "      <td>0.038462</td>\n",
              "      <td>0.320755</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>E1597G</th>\n",
              "      <td>0.066327</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.032609</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.066327</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>K1603T</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.054878</td>\n",
              "      <td>0.056604</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.056604</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>A1746S</th>\n",
              "      <td>0.153061</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.239130</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>I1868T</th>\n",
              "      <td>0.250000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.206522</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.250000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>P1874L</th>\n",
              "      <td>0.224490</td>\n",
              "      <td>0.073171</td>\n",
              "      <td>0.056604</td>\n",
              "      <td>0.260870</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.260870</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>P1874S</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.213415</td>\n",
              "      <td>0.169811</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.213415</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>A1934V</th>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.121951</td>\n",
              "      <td>0.132075</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.132075</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-88a4a758-bb32-4762-b31d-203365b338d0')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-88a4a758-bb32-4762-b31d-203365b338d0 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-88a4a758-bb32-4762-b31d-203365b338d0');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "           frq_BF-09_gamb_2012  ...    max_af\n",
              "aa_change                       ...          \n",
              "V402L                 0.000000  ...  0.121951\n",
              "T791M                 0.163265  ...  0.239130\n",
              "L995F                 1.000000  ...  1.000000\n",
              "I1527T                0.000000  ...  0.121951\n",
              "N1570Y                0.209184  ...  0.320755\n",
              "E1597G                0.066327  ...  0.066327\n",
              "K1603T                0.000000  ...  0.056604\n",
              "A1746S                0.153061  ...  0.239130\n",
              "I1868T                0.250000  ...  0.250000\n",
              "P1874L                0.224490  ...  0.260870\n",
              "P1874S                0.000000  ...  0.213415\n",
              "A1934V                0.000000  ...  0.132075\n",
              "\n",
              "[12 rows x 6 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 24
        }
      ],
      "source": [
        "aa_filt_df = aa_allele_freqs_df.query(\"max_af >= 0.05\")\n",
        "aa_filt_df"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "FJneZ0gxJUD2",
      "metadata": {
        "id": "FJneZ0gxJUD2"
      },
      "source": [
        "Now we can visualise these frequencies the same way we did before."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "id": "03nN3NAgJZsJ",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 542
        },
        "id": "03nN3NAgJZsJ",
        "outputId": "2bedc2b8-2b68-4e79-b522-02654d3c10ab"
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<html>\n",
              "<head><meta charset=\"utf-8\" /></head>\n",
              "<body>\n",
              "    <div>            <script src=\"https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG\"></script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script>                <script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script>\n",
              "        <script src=\"https://cdn.plot.ly/plotly-2.8.3.min.js\"></script>                <div id=\"2f468535-afef-42e0-a924-93e51f71826f\" class=\"plotly-graph-div\" style=\"height:525px; width:600px;\"></div>            <script type=\"text/javascript\">                                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"2f468535-afef-42e0-a924-93e51f71826f\")) {                    Plotly.newPlot(                        \"2f468535-afef-42e0-a924-93e51f71826f\",                        [{\"coloraxis\":\"coloraxis\",\"name\":\"0\",\"texttemplate\":\"%{z:.0%}\",\"x\":[\"BF-09_gamb_2012\",\"BF-09_colu_2012\",\"BF-09_colu_2014\",\"BF-09_gamb_2014\",\"BF-07_gamb_2004\"],\"y\":[\"V402L\",\"T791M\",\"L995F\",\"I1527T\",\"N1570Y\",\"E1597G\",\"K1603T\",\"A1746S\",\"I1868T\",\"P1874L\",\"P1874S\",\"A1934V\"],\"z\":[[0.0,0.12195121951219512,0.11320754716981132,0.0,0.0],[0.16326530612244897,0.018292682926829267,0.0,0.2391304347826087,0.0],[1.0,0.8658536585365854,0.8867924528301887,1.0,0.07692307692307693],[0.0,0.12195121951219512,0.11320754716981132,0.0,0.0],[0.20918367346938777,0.25,0.32075471698113206,0.14130434782608695,0.038461538461538464],[0.0663265306122449,0.0,0.0,0.03260869565217391,0.0],[0.0,0.054878048780487805,0.05660377358490566,0.0,0.0],[0.15306122448979592,0.0,0.0,0.2391304347826087,0.0],[0.25,0.0,0.0,0.20652173913043478,0.0],[0.22448979591836735,0.07317073170731707,0.05660377358490566,0.2608695652173913,0.0],[0.0,0.21341463414634146,0.16981132075471697,0.0,0.0],[0.0,0.12195121951219512,0.1320754716981132,0.0,0.0]],\"type\":\"heatmap\",\"xaxis\":\"x\",\"yaxis\":\"y\",\"hovertemplate\":\"x: %{x}<br>y: %{y}<br>color: %{z}<extra></extra>\"}],                        {\"template\":{\"data\":{\"bar\":[{\"error_x\":{\"color\":\"#2a3f5f\"},\"error_y\":{\"color\":\"#2a3f5f\"},\"marker\":{\"line\":{\"color\":\"#E5ECF6\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"bar\"}],\"barpolar\":[{\"marker\":{\"line\":{\"color\":\"#E5ECF6\",\"width\":0.5},\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"barpolar\"}],\"carpet\":[{\"aaxis\":{\"endlinecolor\":\"#2a3f5f\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"#2a3f5f\"},\"baxis\":{\"endlinecolor\":\"#2a3f5f\",\"gridcolor\":\"white\",\"linecolor\":\"white\",\"minorgridcolor\":\"white\",\"startlinecolor\":\"#2a3f5f\"},\"type\":\"carpet\"}],\"choropleth\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"type\":\"choropleth\"}],\"contour\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"contour\"}],\"contourcarpet\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"type\":\"contourcarpet\"}],\"heatmap\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"heatmap\"}],\"heatmapgl\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"heatmapgl\"}],\"histogram\":[{\"marker\":{\"pattern\":{\"fillmode\":\"overlay\",\"size\":10,\"solidity\":0.2}},\"type\":\"histogram\"}],\"histogram2d\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"histogram2d\"}],\"histogram2dcontour\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"histogram2dcontour\"}],\"mesh3d\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"type\":\"mesh3d\"}],\"parcoords\":[{\"line\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"parcoords\"}],\"pie\":[{\"automargin\":true,\"type\":\"pie\"}],\"scatter\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatter\"}],\"scatter3d\":[{\"line\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatter3d\"}],\"scattercarpet\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattercarpet\"}],\"scattergeo\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattergeo\"}],\"scattergl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattergl\"}],\"scattermapbox\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scattermapbox\"}],\"scatterpolar\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatterpolar\"}],\"scatterpolargl\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatterpolargl\"}],\"scatterternary\":[{\"marker\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"type\":\"scatterternary\"}],\"surface\":[{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"},\"colorscale\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"type\":\"surface\"}],\"table\":[{\"cells\":{\"fill\":{\"color\":\"#EBF0F8\"},\"line\":{\"color\":\"white\"}},\"header\":{\"fill\":{\"color\":\"#C8D4E3\"},\"line\":{\"color\":\"white\"}},\"type\":\"table\"}]},\"layout\":{\"annotationdefaults\":{\"arrowcolor\":\"#2a3f5f\",\"arrowhead\":0,\"arrowwidth\":1},\"autotypenumbers\":\"strict\",\"coloraxis\":{\"colorbar\":{\"outlinewidth\":0,\"ticks\":\"\"}},\"colorscale\":{\"diverging\":[[0,\"#8e0152\"],[0.1,\"#c51b7d\"],[0.2,\"#de77ae\"],[0.3,\"#f1b6da\"],[0.4,\"#fde0ef\"],[0.5,\"#f7f7f7\"],[0.6,\"#e6f5d0\"],[0.7,\"#b8e186\"],[0.8,\"#7fbc41\"],[0.9,\"#4d9221\"],[1,\"#276419\"]],\"sequential\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]],\"sequentialminus\":[[0.0,\"#0d0887\"],[0.1111111111111111,\"#46039f\"],[0.2222222222222222,\"#7201a8\"],[0.3333333333333333,\"#9c179e\"],[0.4444444444444444,\"#bd3786\"],[0.5555555555555556,\"#d8576b\"],[0.6666666666666666,\"#ed7953\"],[0.7777777777777778,\"#fb9f3a\"],[0.8888888888888888,\"#fdca26\"],[1.0,\"#f0f921\"]]},\"colorway\":[\"#636efa\",\"#EF553B\",\"#00cc96\",\"#ab63fa\",\"#FFA15A\",\"#19d3f3\",\"#FF6692\",\"#B6E880\",\"#FF97FF\",\"#FECB52\"],\"font\":{\"color\":\"#2a3f5f\"},\"geo\":{\"bgcolor\":\"white\",\"lakecolor\":\"white\",\"landcolor\":\"#E5ECF6\",\"showlakes\":true,\"showland\":true,\"subunitcolor\":\"white\"},\"hoverlabel\":{\"align\":\"left\"},\"hovermode\":\"closest\",\"mapbox\":{\"style\":\"light\"},\"paper_bgcolor\":\"white\",\"plot_bgcolor\":\"#E5ECF6\",\"polar\":{\"angularaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"},\"bgcolor\":\"#E5ECF6\",\"radialaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"}},\"scene\":{\"xaxis\":{\"backgroundcolor\":\"#E5ECF6\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"},\"yaxis\":{\"backgroundcolor\":\"#E5ECF6\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"},\"zaxis\":{\"backgroundcolor\":\"#E5ECF6\",\"gridcolor\":\"white\",\"gridwidth\":2,\"linecolor\":\"white\",\"showbackground\":true,\"ticks\":\"\",\"zerolinecolor\":\"white\"}},\"shapedefaults\":{\"line\":{\"color\":\"#2a3f5f\"}},\"ternary\":{\"aaxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"},\"baxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"},\"bgcolor\":\"#E5ECF6\",\"caxis\":{\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\"}},\"title\":{\"x\":0.05},\"xaxis\":{\"automargin\":true,\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\",\"title\":{\"standoff\":15},\"zerolinecolor\":\"white\",\"zerolinewidth\":2},\"yaxis\":{\"automargin\":true,\"gridcolor\":\"white\",\"linecolor\":\"white\",\"ticks\":\"\",\"title\":{\"standoff\":15},\"zerolinecolor\":\"white\",\"zerolinewidth\":2}}},\"xaxis\":{\"anchor\":\"y\",\"domain\":[0.0,1.0],\"side\":\"top\",\"tickangle\":-90,\"title\":{\"text\":\"cohorts\"}},\"yaxis\":{\"anchor\":\"x\",\"domain\":[0.0,1.0],\"autorange\":\"reversed\"},\"coloraxis\":{\"colorscale\":[[0.0,\"rgb(255,245,240)\"],[0.125,\"rgb(254,224,210)\"],[0.25,\"rgb(252,187,161)\"],[0.375,\"rgb(252,146,114)\"],[0.5,\"rgb(251,106,74)\"],[0.625,\"rgb(239,59,44)\"],[0.75,\"rgb(203,24,29)\"],[0.875,\"rgb(165,15,21)\"],[1.0,\"rgb(103,0,13)\"]],\"cmin\":0,\"cmax\":1,\"colorbar\":{\"title\":{\"text\":\"frequency\"},\"tickvals\":[0,0.2,0.4,0.6,0.8,1.0],\"ticktext\":[\"0%\",\"20%\",\"40%\",\"60%\",\"80%\",\"100%\"]}},\"margin\":{\"t\":60},\"width\":600},                        {\"responsive\": true}                    ).then(function(){\n",
              "                            \n",
              "var gd = document.getElementById('2f468535-afef-42e0-a924-93e51f71826f');\n",
              "var x = new MutationObserver(function (mutations, observer) {{\n",
              "        var display = window.getComputedStyle(gd).display;\n",
              "        if (!display || display === 'none') {{\n",
              "            console.log([gd, 'removed!']);\n",
              "            Plotly.purge(gd);\n",
              "            observer.disconnect();\n",
              "        }}\n",
              "}});\n",
              "\n",
              "// Listen for the removal of the full notebook cells\n",
              "var notebookContainer = gd.closest('#notebook-container');\n",
              "if (notebookContainer) {{\n",
              "    x.observe(notebookContainer, {childList: true});\n",
              "}}\n",
              "\n",
              "// Listen for the clearing of the current output cell\n",
              "var outputEl = gd.closest('.output');\n",
              "if (outputEl) {{\n",
              "    x.observe(outputEl, {childList: true});\n",
              "}}\n",
              "\n",
              "                        })                };                            </script>        </div>\n",
              "</body>\n",
              "</html>"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "ag3.plot_frequencies_heatmap(aa_filt_df, width=600)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "CLQtluREJbCx",
      "metadata": {
        "id": "CLQtluREJbCx"
      },
      "source": [
        "## Well done!\n",
        "\n",
        "In this module we have learnt how to analyse SNP mutations in the target of pyrethroid insecticides, the voltage-gated sodium-channel. We have calculated the allele frequencies of the SNPs in cohorts of mosquitoes and learnt how to filter and plot them for ease of interpretation."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4629c71d-6de8-4514-a8c0-39742ba54c22",
      "metadata": {
        "id": "4629c71d-6de8-4514-a8c0-39742ba54c22"
      },
      "source": [
        "## Practical exercises\n",
        "\n",
        "### English\n",
        "\n",
        "1. Open this notebook in Google Colab and run it for yourself from top to bottom. Hint: click the rocket icon at the top of the page, then select ‚ÄúColab‚Äù from the drop-down menu.\n",
        "2. Looking at the heatmap output (either amino acid or SNP), can you spot a relationship between the `V402L` and `L995F` frequencies? If so, what is it?\n",
        "2. Re-run the whole analysis but using the Ghanaian sample set. Hint: Try `sample_sets = \"AG1000G-GH\"`. Or any other samples of interest.\n",
        "3. What are the cohorts for this new sample set? Hint: see `frequency_columns`.\n",
        "4. Above, we looked at the _kdr_ \"West\" SNP position, compare and contrast this with the _kdr_ \"East\" SNP position. What is the amino acid change for _kdr_ \"East\"? Hint: The position is 2422651. \n",
        "5. For the Ghanaian sample set, add the a y_label which says \"aa change\" to the amino acid frequency heatmap. Hint: you can view all the function's parameters with `ag3.aa_allele_frequencies?`.\n",
        "6. Remove the colorbar from the same heatmap. Hint: `False`.\n",
        "7. Is the same relationship between `V402L` and `L995F` frequencies present in Ghana? What might be an evolutionary interpretation of this relationship?\n",
        "\n",
        "**When you‚Äôve had enough, create a link to your notebook, and share it with someone. If you‚Äôre attending a training workshop, paste the link into the workshop slack channel.**\n",
        "\n",
        "### Fran√ßais\n",
        "\n",
        "Coming soon."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [],
      "name": "Workshop 1 - Module 4 - Analysing SNPs in the Vgsc gene - 2022-02-14.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.12"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}