{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "628dbc24-27a8-44e8-a58e-a6ae3ed09366",
   "metadata": {
    "tags": []
   },
   "source": [
    "![banner](../banner.jpg)\n",
    "\n",
    "***Training course in data analysis for genomic surveillance of African malaria vectors***\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7475bf01-93e5-4e35-b1b3-9f7a638194fb",
   "metadata": {},
   "source": [
    "# Exploration of the _Anopheles funestus_ resource\n",
    "\n",
    "**Theme: Data**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4344e1f6-4c7e-427f-99ed-cf245bd26932",
   "metadata": {},
   "source": [
    "This module provides an introduction to accessing and exploring data about *Anopheles funestus* mosquito specimens collected in the field and submitted for whole-genome sequencing by MalariaGEN."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c178fc5-1f5e-46f2-b9d4-7e8aef6d3373",
   "metadata": {},
   "source": [
    "A certain degree of familiarity with the content of the training course on _Anopheles gambiae s.l._ is expected. Many of the concepts presented in this module were introduced in the training course and we will refer to the relevant workshop and module instead of giving detailed explanations."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94eae6a7-a568-4a75-9de2-18d07dc99a7e",
   "metadata": {},
   "source": [
    "## Learning objectives\n",
    "\n",
    "After completing this module, you will be able to:\n",
    "\n",
    "* Use the `malariagen_data` Python package to access `Af1.0` data in Google Cloud.\n",
    "* Explore the `Af1.0` data release and summarise the mosquito samples for which genomic data are available using pivot tables and maps.\n",
    "* Use genomic coordinates to identify sites and regions within the reference genome.\n",
    "* Find specific genes of interest in the reference genome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37231d05-ade6-4bb6-a1a6-f227be9b9174",
   "metadata": {},
   "source": [
    "## Lecture?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b53e6d3e-14ff-4003-84ce-bf3b56162110",
   "metadata": {},
   "source": [
    "### English"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3fe3931-8590-4238-a05b-248b15ba657f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6756c1dd-eaf5-4ad7-bd96-c4f9c3c17033",
   "metadata": {},
   "source": [
    "### Fran√ßais"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f33a947b-80a3-459d-9beb-8cb19bd6ab28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d666863b-b9df-4428-924d-b19f53ee01a1",
   "metadata": {
    "tags": []
   },
   "source": [
    "Please note that the code in the cells below might differ from that shown in the video. This can happen because Python packages and their dependencies change due to updates, necessitating tweaks to the code."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26cccc59-b417-45d7-8e49-0ed77514b467",
   "metadata": {},
   "source": [
    "## Why are we interested in _An. funestus_ ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9ec1496f-7ade-4664-b3af-f68969f9fec0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<img width=\"50%\" height=\"50%\" src=\"https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/src/img/advanced/MAP_Africa_vector_range.png?raw=1\"/>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<img width=\"50%\" height=\"50%\" src=\"https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/src/img/advanced/MAP_Africa_vector_range.png?raw=1\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f06977b6-dd31-4626-816e-2e5cd0d857b3",
   "metadata": {},
   "source": [
    "An. funestus is a major vector of malaria in sub-Saharan Africa, particularly in East Africa. Its range covers most of the continent, as shown on the map above courtesy of the Malaria Atlas Project [1]. If vector control interventions are successful in reducing the populations of _An. gambiae s.l._ withour affecting as much the _An. funestus_ populations, they could lead to _An. funestus_ replacing _An. gambiae s.l._ as a malaria vector. This has, for instance, already been observed in East Africa. Msugupakulya et al. [2] have shown that, while _An. gambiae s.l._ was the main vector of malaria in East Africa during the first decade of the XXI<sup>st</sup> century, _An. funestus_ has taken that role more recently."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c505b271-05ff-4d19-9c07-576625aef8a1",
   "metadata": {},
   "source": [
    "_An. funestus_ is known to be highly anthropophilic [3]. It also been shown to have a significantly longer lifespan than other African malaria vectors [4]. In some areas, it is also reported to have very high Plasmodium infection rates [5, 6, 7]. _An funestus_ has adapted recently to the use of indoor-based interventions, including bed nets, by modifying its behaviour. It is now known to be probe to outdoor, early evening and late morning biting [8, 9, 10]. There is also evidence for a preference for larval habitats that persist in the dry months [11] which may extend the transmission season. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "260b69d1-5a70-4e24-93d0-418a66d2f74f",
   "metadata": {},
   "source": [
    "## Why is this a different course from the one on _An. gambiae s.l._?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82879f29-14f6-4dc6-911e-1ab569046d09",
   "metadata": {},
   "source": [
    "All the methods that were introduced in the _An. gambiae s.l._ course to analyse population structure ([Workshop 3](https://anopheles-genomic-surveillance.github.io/workshop-3/about.html)), genetic diversity ([Workshop 5](https://anopheles-genomic-surveillance.github.io/workshop-5/about.html)), positive selection ([Workshop 6](https://anopheles-genomic-surveillance.github.io/workshop-6/about.html)), gene flow ([Workshop 7](https://anopheles-genomic-surveillance.github.io/workshop-7/about.html)) or insecticide resistance ([Workshop 1](https://anopheles-genomic-surveillance.github.io/workshop-1/about.html) and [Workshop 2](https://anopheles-genomic-surveillance.github.io/workshop-2/about.html)) can be applied to _An. funestus_ (and will be in [Advanced funestus - Analyses](https://anopheles-genomic-surveillance.github.io/advanced-training-materials/funestus_analyses.html)) but, particularly for insecticide resistance analyses, the fact that _An. gambiae s.l._ and _An. funestus_ are two very different species cannot be ignored."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e81e9f58-86ff-4d4a-a528-0b086959529d",
   "metadata": {},
   "source": [
    "To begin with, there is no gene flow between _An. gambiae s.l._ and _An. funestus_. The two species have common ancestors, they do belong to the same genus _Anopheles_, but they diverged around 9.5 million years ago [12]. This, obviously, predates any human intervention as it predates humanity [13]. As a result, any insecticide resistance mutation that is observed will either be different or a co-evolution."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dfbcd85-2122-4073-b5b7-827c965dc4d6",
   "metadata": {},
   "source": [
    "The `Af1.0` resource is built using the `idAnoFuneDA-416_04` [14] genome assembly. As we will see, while the _An. funestus_ genome has the same number of chromosomes (2 autosomes and X) as the _An. gambiae s.l._, they are of different lengths, the inversions are different and the genes of interest may be located in completely different places in the genome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96d3f282-c3ac-487f-ab5d-63dc8f6d6d4b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<img width=\"50%\" height=\"50%\" src=\"https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/src/img/advanced/funestus_synteny.png?raw=1\"/>\n",
       "#I need to find a good picture to put here\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<img width=\"50%\" height=\"50%\" src=\"https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/src/img/advanced/funestus_synteny.png?raw=1\"/>\n",
    "#I need to find a good picture to put here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ab8eab-2594-4176-a0e4-949422c34ac6",
   "metadata": {},
   "source": [
    "## Accessing the `Af1.0` data resource\n",
    "\n",
    "We are now going to be accessing and analysing data from the [*Anopheles funestus* 1000 Genomes Project phase 1 data resource](https://malariagen.github.io/vector-data/af1/af1.0.html), also known as \"Af1.0\" for short. This includes data from whole-genome sequencing of 656 mosquitoes from 13 African countries.\n",
    "\n",
    "To set up your notebook to access these data, first install the [malariagen_data](https://github.com/malariagen/malariagen-data-python) package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd9ae947-e287-443f-b0b9-28e9abadc5dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install -q --no-warn-conflicts malariagen_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc771a4-6498-4878-b2bb-aecadcac031f",
   "metadata": {},
   "source": [
    "We will also need to use ipyleaflet to make some maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "30d0c480-6d18-49b1-9989-f913cc906c98",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install -qq ipyleaflet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "af392f0b-2ec7-474a-b121-fe9e995cb4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "import ipyleaflet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a969d70f-8910-402e-ab3f-6d193ec82818",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "'use strict';\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    function drop(id) {\n",
       "      const view = Bokeh.index.get_by_id(id)\n",
       "      if (view != null) {\n",
       "        view.model.document.clear()\n",
       "        Bokeh.index.delete(view)\n",
       "      }\n",
       "    }\n",
       "\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null) {\n",
       "      drop(id)\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim()\n",
       "            drop(id)\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded(error = null) {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      const html = (() => {\n",
       "        if (typeof root.Bokeh === \"undefined\") {\n",
       "          if (error == null) {\n",
       "            return \"BokehJS is loading ...\";\n",
       "          } else {\n",
       "            return \"BokehJS failed to load.\";\n",
       "          }\n",
       "        } else {\n",
       "          const prefix = `BokehJS ${root.Bokeh.version}`;\n",
       "          if (error == null) {\n",
       "            return `${prefix} successfully loaded.`;\n",
       "          } else {\n",
       "            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n",
       "          }\n",
       "        }\n",
       "      })();\n",
       "      el.innerHTML = html;\n",
       "\n",
       "      if (error != null) {\n",
       "        const wrapper = document.createElement(\"div\");\n",
       "        wrapper.style.overflow = \"auto\";\n",
       "        wrapper.style.height = \"5em\";\n",
       "        wrapper.style.resize = \"vertical\";\n",
       "        const content = document.createElement(\"div\");\n",
       "        content.style.fontFamily = \"monospace\";\n",
       "        content.style.whiteSpace = \"pre-wrap\";\n",
       "        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n",
       "        content.textContent = error.stack ?? error.toString();\n",
       "        wrapper.append(content);\n",
       "        el.append(wrapper);\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(() => display_loaded(error), 100);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.6.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      try {\n",
       "            for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "\n",
       "      } catch (error) {throw error;\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "'use strict';\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded(error = null) {\n    const el = document.getElementById(null);\n    if (el != null) {\n      const html = (() => {\n        if (typeof root.Bokeh === \"undefined\") {\n          if (error == null) {\n            return \"BokehJS is loading ...\";\n          } else {\n            return \"BokehJS failed to load.\";\n          }\n        } else {\n          const prefix = `BokehJS ${root.Bokeh.version}`;\n          if (error == null) {\n            return `${prefix} successfully loaded.`;\n          } else {\n            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n          }\n        }\n      })();\n      el.innerHTML = html;\n\n      if (error != null) {\n        const wrapper = document.createElement(\"div\");\n        wrapper.style.overflow = \"auto\";\n        wrapper.style.height = \"5em\";\n        wrapper.style.resize = \"vertical\";\n        const content = document.createElement(\"div\");\n        content.style.fontFamily = \"monospace\";\n        content.style.whiteSpace = \"pre-wrap\";\n        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n        content.textContent = error.stack ?? error.toString();\n        wrapper.append(content);\n        el.append(wrapper);\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(() => display_loaded(error), 100);\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.6.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n      try {\n            for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n\n      } catch (error) {throw error;\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <table class=\"malariagen-af1\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\" colspan=\"2\">MalariaGEN Af1 API client</th>\n",
       "                    </tr>\n",
       "                    <tr><td colspan=\"2\" style=\"text-align: left\">\n",
       "                        Please note that data are subject to terms of use,\n",
       "                        for more information see <a href=\"https://www.malariagen.net/data\">\n",
       "                        the MalariaGEN website</a> or contact support@malariagen.net.\n",
       "                        See also the <a href=\"https://malariagen.github.io/malariagen-data-python/v15.2.0/Af1.html\">Af1 API docs</a>.\n",
       "                    </td></tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Storage URL\n",
       "                        </th>\n",
       "                        <td>gs://vo_afun_release_master_us_central1</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Data releases available\n",
       "                        </th>\n",
       "                        <td>1.0, 1.1, 1.2, 1.3, 1.4</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Results cache\n",
       "                        </th>\n",
       "                        <td>None</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Cohorts analysis\n",
       "                        </th>\n",
       "                        <td>20240515</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Site filters analysis\n",
       "                        </th>\n",
       "                        <td>dt_20200416</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Software version\n",
       "                        </th>\n",
       "                        <td>malariagen_data 15.2.0</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Client location\n",
       "                        </th>\n",
       "                        <td>Iowa, United States (Google Cloud us-central1)</td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        "
      ],
      "text/plain": [
       "<MalariaGEN Af1 API client>\n",
       "Storage URL             : gs://vo_afun_release_master_us_central1\n",
       "Data releases available : 1.0, 1.1, 1.2, 1.3, 1.4\n",
       "Results cache           : None\n",
       "Cohorts analysis        : 20240515\n",
       "Site filters analysis   : dt_20200416\n",
       "Software version        : malariagen_data 15.2.0\n",
       "Client location         : Iowa, United States (Google Cloud us-central1)\n",
       "---\n",
       "Please note that data are subject to terms of use,\n",
       "for more information see https://www.malariagen.net/data\n",
       "or contact support@malariagen.net. For API documentation see \n",
       "https://malariagen.github.io/malariagen-data-python/v15.2.0/Af1.html"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1 = malariagen_data.Af1()\n",
    "af1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6127057e-80e8-4d73-b6c3-732b3d36b1ea",
   "metadata": {},
   "source": [
    "## Accessing and exploring *Anopheles funestus* genomic data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed53746d-f82e-4c27-9ce5-9e396e90e4e2",
   "metadata": {},
   "source": [
    "### Loading sample metadata\n",
    "\n",
    "We can use the [`sample_metadata()`](https://malariagen.github.io/malariagen-data-python/latest/Af1.html#sample-metadata-access) function to retrieve a pandas DataFrame containing metadata about all 656 samples in the Af1.0 resource. In this DataFrame, each row represents one mosquito sample, and the columns such as `country` and `year` provide information about where the mosquito was originally collected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "77f26ca1-cb5a-4dc8-a336-590978735897",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>contributor</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>sex_call</th>\n",
       "      <th>...</th>\n",
       "      <th>admin1_name</th>\n",
       "      <th>admin1_iso</th>\n",
       "      <th>admin2_name</th>\n",
       "      <th>taxon</th>\n",
       "      <th>cohort_admin1_year</th>\n",
       "      <th>cohort_admin1_month</th>\n",
       "      <th>cohort_admin1_quarter</th>\n",
       "      <th>cohort_admin2_year</th>\n",
       "      <th>cohort_admin2_month</th>\n",
       "      <th>cohort_admin2_quarter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VBS24195</td>\n",
       "      <td>1229-GH-A-GH01</td>\n",
       "      <td>Samuel Dadzie</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Dimabi</td>\n",
       "      <td>2017</td>\n",
       "      <td>8</td>\n",
       "      <td>9.420</td>\n",
       "      <td>-1.083</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Northern Region</td>\n",
       "      <td>GH-NP</td>\n",
       "      <td>Tolon</td>\n",
       "      <td>funestus</td>\n",
       "      <td>GH-NP_fune_2017</td>\n",
       "      <td>GH-NP_fune_2017_08</td>\n",
       "      <td>GH-NP_fune_2017_Q3</td>\n",
       "      <td>GH-NP_Tolon_fune_2017</td>\n",
       "      <td>GH-NP_Tolon_fune_2017_08</td>\n",
       "      <td>GH-NP_Tolon_fune_2017_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VBS24196</td>\n",
       "      <td>1229-GH-A-GH02</td>\n",
       "      <td>Samuel Dadzie</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Gbullung</td>\n",
       "      <td>2017</td>\n",
       "      <td>7</td>\n",
       "      <td>9.488</td>\n",
       "      <td>-1.009</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Northern Region</td>\n",
       "      <td>GH-NP</td>\n",
       "      <td>Kumbungu</td>\n",
       "      <td>funestus</td>\n",
       "      <td>GH-NP_fune_2017</td>\n",
       "      <td>GH-NP_fune_2017_07</td>\n",
       "      <td>GH-NP_fune_2017_Q3</td>\n",
       "      <td>GH-NP_Kumbungu_fune_2017</td>\n",
       "      <td>GH-NP_Kumbungu_fune_2017_07</td>\n",
       "      <td>GH-NP_Kumbungu_fune_2017_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VBS24197</td>\n",
       "      <td>1229-GH-A-GH03</td>\n",
       "      <td>Samuel Dadzie</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Dimabi</td>\n",
       "      <td>2017</td>\n",
       "      <td>7</td>\n",
       "      <td>9.420</td>\n",
       "      <td>-1.083</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Northern Region</td>\n",
       "      <td>GH-NP</td>\n",
       "      <td>Tolon</td>\n",
       "      <td>funestus</td>\n",
       "      <td>GH-NP_fune_2017</td>\n",
       "      <td>GH-NP_fune_2017_07</td>\n",
       "      <td>GH-NP_fune_2017_Q3</td>\n",
       "      <td>GH-NP_Tolon_fune_2017</td>\n",
       "      <td>GH-NP_Tolon_fune_2017_07</td>\n",
       "      <td>GH-NP_Tolon_fune_2017_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VBS24198</td>\n",
       "      <td>1229-GH-A-GH04</td>\n",
       "      <td>Samuel Dadzie</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Dimabi</td>\n",
       "      <td>2017</td>\n",
       "      <td>8</td>\n",
       "      <td>9.420</td>\n",
       "      <td>-1.083</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Northern Region</td>\n",
       "      <td>GH-NP</td>\n",
       "      <td>Tolon</td>\n",
       "      <td>funestus</td>\n",
       "      <td>GH-NP_fune_2017</td>\n",
       "      <td>GH-NP_fune_2017_08</td>\n",
       "      <td>GH-NP_fune_2017_Q3</td>\n",
       "      <td>GH-NP_Tolon_fune_2017</td>\n",
       "      <td>GH-NP_Tolon_fune_2017_08</td>\n",
       "      <td>GH-NP_Tolon_fune_2017_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VBS24199</td>\n",
       "      <td>1229-GH-A-GH05</td>\n",
       "      <td>Samuel Dadzie</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Gupanarigu</td>\n",
       "      <td>2017</td>\n",
       "      <td>8</td>\n",
       "      <td>9.497</td>\n",
       "      <td>-0.952</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Northern Region</td>\n",
       "      <td>GH-NP</td>\n",
       "      <td>Kumbungu</td>\n",
       "      <td>funestus</td>\n",
       "      <td>GH-NP_fune_2017</td>\n",
       "      <td>GH-NP_fune_2017_08</td>\n",
       "      <td>GH-NP_fune_2017_Q3</td>\n",
       "      <td>GH-NP_Kumbungu_fune_2017</td>\n",
       "      <td>GH-NP_Kumbungu_fune_2017_08</td>\n",
       "      <td>GH-NP_Kumbungu_fune_2017_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>VBS24534</td>\n",
       "      <td>1240-MZ-A-MozF_1314</td>\n",
       "      <td>Lizette Koekemoer</td>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Motinho</td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>-10.851</td>\n",
       "      <td>40.594</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cabo Delgado</td>\n",
       "      <td>MZ-P</td>\n",
       "      <td>Palma</td>\n",
       "      <td>funestus</td>\n",
       "      <td>MZ-P_fune_2015</td>\n",
       "      <td>MZ-P_fune_2015_08</td>\n",
       "      <td>MZ-P_fune_2015_Q3</td>\n",
       "      <td>MZ-P_Palma_fune_2015</td>\n",
       "      <td>MZ-P_Palma_fune_2015_08</td>\n",
       "      <td>MZ-P_Palma_fune_2015_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>VBS24535</td>\n",
       "      <td>1240-MZ-A-MozF_1315</td>\n",
       "      <td>Lizette Koekemoer</td>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Motinho</td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>-10.851</td>\n",
       "      <td>40.594</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cabo Delgado</td>\n",
       "      <td>MZ-P</td>\n",
       "      <td>Palma</td>\n",
       "      <td>funestus</td>\n",
       "      <td>MZ-P_fune_2015</td>\n",
       "      <td>MZ-P_fune_2015_08</td>\n",
       "      <td>MZ-P_fune_2015_Q3</td>\n",
       "      <td>MZ-P_Palma_fune_2015</td>\n",
       "      <td>MZ-P_Palma_fune_2015_08</td>\n",
       "      <td>MZ-P_Palma_fune_2015_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>VBS24536</td>\n",
       "      <td>1240-MZ-A-MozF_1317</td>\n",
       "      <td>Lizette Koekemoer</td>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Motinho</td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>-10.851</td>\n",
       "      <td>40.594</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cabo Delgado</td>\n",
       "      <td>MZ-P</td>\n",
       "      <td>Palma</td>\n",
       "      <td>funestus</td>\n",
       "      <td>MZ-P_fune_2015</td>\n",
       "      <td>MZ-P_fune_2015_08</td>\n",
       "      <td>MZ-P_fune_2015_Q3</td>\n",
       "      <td>MZ-P_Palma_fune_2015</td>\n",
       "      <td>MZ-P_Palma_fune_2015_08</td>\n",
       "      <td>MZ-P_Palma_fune_2015_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>VBS24537</td>\n",
       "      <td>1240-MZ-A-MozF_1319</td>\n",
       "      <td>Lizette Koekemoer</td>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Motinho</td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>-10.851</td>\n",
       "      <td>40.594</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cabo Delgado</td>\n",
       "      <td>MZ-P</td>\n",
       "      <td>Palma</td>\n",
       "      <td>funestus</td>\n",
       "      <td>MZ-P_fune_2015</td>\n",
       "      <td>MZ-P_fune_2015_08</td>\n",
       "      <td>MZ-P_fune_2015_Q3</td>\n",
       "      <td>MZ-P_Palma_fune_2015</td>\n",
       "      <td>MZ-P_Palma_fune_2015_08</td>\n",
       "      <td>MZ-P_Palma_fune_2015_Q3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655</th>\n",
       "      <td>VBS24539</td>\n",
       "      <td>1240-MZ-A-MozF_1323</td>\n",
       "      <td>Lizette Koekemoer</td>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Motinho</td>\n",
       "      <td>2015</td>\n",
       "      <td>8</td>\n",
       "      <td>-10.851</td>\n",
       "      <td>40.594</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>Cabo Delgado</td>\n",
       "      <td>MZ-P</td>\n",
       "      <td>Palma</td>\n",
       "      <td>funestus</td>\n",
       "      <td>MZ-P_fune_2015</td>\n",
       "      <td>MZ-P_fune_2015_08</td>\n",
       "      <td>MZ-P_fune_2015_Q3</td>\n",
       "      <td>MZ-P_Palma_fune_2015</td>\n",
       "      <td>MZ-P_Palma_fune_2015_08</td>\n",
       "      <td>MZ-P_Palma_fune_2015_Q3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>656 rows √ó 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_id    partner_sample_id        contributor     country    location  \\\n",
       "0    VBS24195       1229-GH-A-GH01      Samuel Dadzie       Ghana      Dimabi   \n",
       "1    VBS24196       1229-GH-A-GH02      Samuel Dadzie       Ghana    Gbullung   \n",
       "2    VBS24197       1229-GH-A-GH03      Samuel Dadzie       Ghana      Dimabi   \n",
       "3    VBS24198       1229-GH-A-GH04      Samuel Dadzie       Ghana      Dimabi   \n",
       "4    VBS24199       1229-GH-A-GH05      Samuel Dadzie       Ghana  Gupanarigu   \n",
       "..        ...                  ...                ...         ...         ...   \n",
       "651  VBS24534  1240-MZ-A-MozF_1314  Lizette Koekemoer  Mozambique     Motinho   \n",
       "652  VBS24535  1240-MZ-A-MozF_1315  Lizette Koekemoer  Mozambique     Motinho   \n",
       "653  VBS24536  1240-MZ-A-MozF_1317  Lizette Koekemoer  Mozambique     Motinho   \n",
       "654  VBS24537  1240-MZ-A-MozF_1319  Lizette Koekemoer  Mozambique     Motinho   \n",
       "655  VBS24539  1240-MZ-A-MozF_1323  Lizette Koekemoer  Mozambique     Motinho   \n",
       "\n",
       "     year  month  latitude  longitude sex_call  ...      admin1_name  \\\n",
       "0    2017      8     9.420     -1.083        F  ...  Northern Region   \n",
       "1    2017      7     9.488     -1.009        F  ...  Northern Region   \n",
       "2    2017      7     9.420     -1.083        F  ...  Northern Region   \n",
       "3    2017      8     9.420     -1.083        F  ...  Northern Region   \n",
       "4    2017      8     9.497     -0.952        F  ...  Northern Region   \n",
       "..    ...    ...       ...        ...      ...  ...              ...   \n",
       "651  2015      8   -10.851     40.594        F  ...     Cabo Delgado   \n",
       "652  2015      8   -10.851     40.594        F  ...     Cabo Delgado   \n",
       "653  2015      8   -10.851     40.594        F  ...     Cabo Delgado   \n",
       "654  2015      8   -10.851     40.594        F  ...     Cabo Delgado   \n",
       "655  2015      8   -10.851     40.594        F  ...     Cabo Delgado   \n",
       "\n",
       "    admin1_iso  admin2_name     taxon cohort_admin1_year cohort_admin1_month  \\\n",
       "0        GH-NP        Tolon  funestus    GH-NP_fune_2017  GH-NP_fune_2017_08   \n",
       "1        GH-NP     Kumbungu  funestus    GH-NP_fune_2017  GH-NP_fune_2017_07   \n",
       "2        GH-NP        Tolon  funestus    GH-NP_fune_2017  GH-NP_fune_2017_07   \n",
       "3        GH-NP        Tolon  funestus    GH-NP_fune_2017  GH-NP_fune_2017_08   \n",
       "4        GH-NP     Kumbungu  funestus    GH-NP_fune_2017  GH-NP_fune_2017_08   \n",
       "..         ...          ...       ...                ...                 ...   \n",
       "651       MZ-P        Palma  funestus     MZ-P_fune_2015   MZ-P_fune_2015_08   \n",
       "652       MZ-P        Palma  funestus     MZ-P_fune_2015   MZ-P_fune_2015_08   \n",
       "653       MZ-P        Palma  funestus     MZ-P_fune_2015   MZ-P_fune_2015_08   \n",
       "654       MZ-P        Palma  funestus     MZ-P_fune_2015   MZ-P_fune_2015_08   \n",
       "655       MZ-P        Palma  funestus     MZ-P_fune_2015   MZ-P_fune_2015_08   \n",
       "\n",
       "    cohort_admin1_quarter        cohort_admin2_year  \\\n",
       "0      GH-NP_fune_2017_Q3     GH-NP_Tolon_fune_2017   \n",
       "1      GH-NP_fune_2017_Q3  GH-NP_Kumbungu_fune_2017   \n",
       "2      GH-NP_fune_2017_Q3     GH-NP_Tolon_fune_2017   \n",
       "3      GH-NP_fune_2017_Q3     GH-NP_Tolon_fune_2017   \n",
       "4      GH-NP_fune_2017_Q3  GH-NP_Kumbungu_fune_2017   \n",
       "..                    ...                       ...   \n",
       "651     MZ-P_fune_2015_Q3      MZ-P_Palma_fune_2015   \n",
       "652     MZ-P_fune_2015_Q3      MZ-P_Palma_fune_2015   \n",
       "653     MZ-P_fune_2015_Q3      MZ-P_Palma_fune_2015   \n",
       "654     MZ-P_fune_2015_Q3      MZ-P_Palma_fune_2015   \n",
       "655     MZ-P_fune_2015_Q3      MZ-P_Palma_fune_2015   \n",
       "\n",
       "             cohort_admin2_month        cohort_admin2_quarter  \n",
       "0       GH-NP_Tolon_fune_2017_08     GH-NP_Tolon_fune_2017_Q3  \n",
       "1    GH-NP_Kumbungu_fune_2017_07  GH-NP_Kumbungu_fune_2017_Q3  \n",
       "2       GH-NP_Tolon_fune_2017_07     GH-NP_Tolon_fune_2017_Q3  \n",
       "3       GH-NP_Tolon_fune_2017_08     GH-NP_Tolon_fune_2017_Q3  \n",
       "4    GH-NP_Kumbungu_fune_2017_08  GH-NP_Kumbungu_fune_2017_Q3  \n",
       "..                           ...                          ...  \n",
       "651      MZ-P_Palma_fune_2015_08      MZ-P_Palma_fune_2015_Q3  \n",
       "652      MZ-P_Palma_fune_2015_08      MZ-P_Palma_fune_2015_Q3  \n",
       "653      MZ-P_Palma_fune_2015_08      MZ-P_Palma_fune_2015_Q3  \n",
       "654      MZ-P_Palma_fune_2015_08      MZ-P_Palma_fune_2015_Q3  \n",
       "655      MZ-P_Palma_fune_2015_08      MZ-P_Palma_fune_2015_Q3  \n",
       "\n",
       "[656 rows x 45 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_samples = af1.sample_metadata(sample_sets=\"1.0\")\n",
    "df_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26e7451f-aa5d-40a5-8b6a-418096248cee",
   "metadata": {},
   "source": [
    "We can then group these samples according to country and year of collection, sample set, and taxon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1c70b9d7-b0cb-4c6e-b383-39d052f3f2f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "country                           year  sample_set                     taxon   \n",
       "Benin                             2014  1231-VO-MULTI-WONDJI-VMF00043  funestus    37\n",
       "Cameroon                          2014  1231-VO-MULTI-WONDJI-VMF00043  funestus    45\n",
       "Central African Republic          2016  1230-VO-GA-CF-AYALA-VMF00045   funestus    10\n",
       "Democratic Republic of the Congo  2015  1231-VO-MULTI-WONDJI-VMF00043  funestus    34\n",
       "                                  2017  1240-VO-CD-KOEKEMOER-VMF00099  funestus    43\n",
       "Gabon                             2017  1230-VO-GA-CF-AYALA-VMF00045   funestus    40\n",
       "Ghana                             2014  1231-VO-MULTI-WONDJI-VMF00043  funestus    31\n",
       "                                  2017  1229-VO-GH-DADZIE-VMF00095     funestus    36\n",
       "Kenya                             2014  1232-VO-KE-OCHOMO-VMF00044     funestus    37\n",
       "                                  2016  1232-VO-KE-OCHOMO-VMF00044     funestus    44\n",
       "Malawi                            2014  1231-VO-MULTI-WONDJI-VMF00043  funestus    18\n",
       "Mozambique                        2015  1240-VO-MZ-KOEKEMOER-VMF00101  funestus    40\n",
       "                                  2016  1231-VO-MULTI-WONDJI-VMF00043  funestus    22\n",
       "                                  2018  1235-VO-MZ-PAAIJMANS-VMF00094  funestus    76\n",
       "Nigeria                           2015  1231-VO-MULTI-WONDJI-VMF00043  funestus    41\n",
       "Tanzania                          2017  1236-VO-TZ-OKUMU-VMF00090      funestus    10\n",
       "Uganda                            2014  1231-VO-MULTI-WONDJI-VMF00043  funestus    49\n",
       "Zambia                            2016  1231-VO-MULTI-WONDJI-VMF00043  funestus    43\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_samples.groupby([\"country\", \"year\", \"sample_set\", \"taxon\"]).size()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3d2fbad-6219-4d1c-81a3-c7adbcf6a046",
   "metadata": {},
   "source": [
    "As we can see, all samples were collected between 2014 and 2018 and all samples were identified as _An. funestus_. We will thus not be able to look at other members of the complex (e.g., _An. vaneedeni_). We can also see that most countries are only represented by one sample set."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f35bdc4-2b72-451e-87d9-24e0730d28aa",
   "metadata": {},
   "source": [
    "We can represent the provenance of the samples in a pivot table to better see the temporal distribution of samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86e45e03-c03a-4ffc-a176-28a4734a954d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>year</th>\n",
       "      <th>2014</th>\n",
       "      <th>2015</th>\n",
       "      <th>2016</th>\n",
       "      <th>2017</th>\n",
       "      <th>2018</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Benin</th>\n",
       "      <th>Kpome</th>\n",
       "      <th>6.383</th>\n",
       "      <th>2.200</th>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cameroon</th>\n",
       "      <th>Mibellon</th>\n",
       "      <th>6.083</th>\n",
       "      <th>11.500</th>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Central African Republic</th>\n",
       "      <th>Bangui</th>\n",
       "      <th>4.377</th>\n",
       "      <th>18.608</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Democratic Republic of the Congo</th>\n",
       "      <th>Kinshasa</th>\n",
       "      <th>-4.327</th>\n",
       "      <th>15.313</th>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Watsa Moke</th>\n",
       "      <th>3.094</th>\n",
       "      <th>29.548</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gabon</th>\n",
       "      <th>Benguia</th>\n",
       "      <th>-1.633</th>\n",
       "      <th>13.492</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Ghana</th>\n",
       "      <th>Dimabi</th>\n",
       "      <th>9.420</th>\n",
       "      <th>-1.083</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gbullung</th>\n",
       "      <th>9.488</th>\n",
       "      <th>-1.009</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gupanarigu</th>\n",
       "      <th>9.497</th>\n",
       "      <th>-0.952</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Obuasi</th>\n",
       "      <th>5.933</th>\n",
       "      <th>-1.617</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Woribogu</th>\n",
       "      <th>8.869</th>\n",
       "      <th>0.321</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Kenya</th>\n",
       "      <th>Ahero</th>\n",
       "      <th>-0.174</th>\n",
       "      <th>34.920</th>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Awendo</th>\n",
       "      <th>-0.901</th>\n",
       "      <th>34.533</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nyamilu</th>\n",
       "      <th>-1.074</th>\n",
       "      <th>34.391</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sumba</th>\n",
       "      <th>-0.686</th>\n",
       "      <th>34.596</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uradi</th>\n",
       "      <th>0.980</th>\n",
       "      <th>34.543</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uriri</th>\n",
       "      <th>-0.952</th>\n",
       "      <th>34.513</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Malawi</th>\n",
       "      <th>Chikwawa</th>\n",
       "      <th>-16.017</th>\n",
       "      <th>34.789</th>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Mozambique</th>\n",
       "      <th>Motinho</th>\n",
       "      <th>-10.851</th>\n",
       "      <th>40.594</th>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Palmeira</th>\n",
       "      <th>-25.255</th>\n",
       "      <th>32.873</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Palmeiras</th>\n",
       "      <th>-25.265</th>\n",
       "      <th>32.877</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nigeria</th>\n",
       "      <th>Akaka Remo</th>\n",
       "      <th>6.950</th>\n",
       "      <th>3.717</th>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tanzania</th>\n",
       "      <th>Minepa Village</th>\n",
       "      <th>-8.273</th>\n",
       "      <th>36.681</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uganda</th>\n",
       "      <th>Tororo</th>\n",
       "      <th>0.750</th>\n",
       "      <th>34.083</th>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zambia</th>\n",
       "      <th>Katete</th>\n",
       "      <th>-14.183</th>\n",
       "      <th>31.867</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "year                                                                2014  \\\n",
       "country                          location       latitude longitude         \n",
       "Benin                            Kpome           6.383    2.200       37   \n",
       "Cameroon                         Mibellon        6.083    11.500      45   \n",
       "Central African Republic         Bangui          4.377    18.608       0   \n",
       "Democratic Republic of the Congo Kinshasa       -4.327    15.313       0   \n",
       "                                 Watsa Moke      3.094    29.548       0   \n",
       "Gabon                            Benguia        -1.633    13.492       0   \n",
       "Ghana                            Dimabi          9.420   -1.083        0   \n",
       "                                 Gbullung        9.488   -1.009        0   \n",
       "                                 Gupanarigu      9.497   -0.952        0   \n",
       "                                 Obuasi          5.933   -1.617       31   \n",
       "                                 Woribogu        8.869    0.321        0   \n",
       "Kenya                            Ahero          -0.174    34.920      37   \n",
       "                                 Awendo         -0.901    34.533       0   \n",
       "                                 Nyamilu        -1.074    34.391       0   \n",
       "                                 Sumba          -0.686    34.596       0   \n",
       "                                 Uradi           0.980    34.543       0   \n",
       "                                 Uriri          -0.952    34.513       0   \n",
       "Malawi                           Chikwawa       -16.017   34.789      18   \n",
       "Mozambique                       Motinho        -10.851   40.594       0   \n",
       "                                 Palmeira       -25.255   32.873       0   \n",
       "                                 Palmeiras      -25.265   32.877       0   \n",
       "Nigeria                          Akaka Remo      6.950    3.717        0   \n",
       "Tanzania                         Minepa Village -8.273    36.681       0   \n",
       "Uganda                           Tororo          0.750    34.083      49   \n",
       "Zambia                           Katete         -14.183   31.867       0   \n",
       "\n",
       "year                                                                2015  \\\n",
       "country                          location       latitude longitude         \n",
       "Benin                            Kpome           6.383    2.200        0   \n",
       "Cameroon                         Mibellon        6.083    11.500       0   \n",
       "Central African Republic         Bangui          4.377    18.608       0   \n",
       "Democratic Republic of the Congo Kinshasa       -4.327    15.313      34   \n",
       "                                 Watsa Moke      3.094    29.548       0   \n",
       "Gabon                            Benguia        -1.633    13.492       0   \n",
       "Ghana                            Dimabi          9.420   -1.083        0   \n",
       "                                 Gbullung        9.488   -1.009        0   \n",
       "                                 Gupanarigu      9.497   -0.952        0   \n",
       "                                 Obuasi          5.933   -1.617        0   \n",
       "                                 Woribogu        8.869    0.321        0   \n",
       "Kenya                            Ahero          -0.174    34.920       0   \n",
       "                                 Awendo         -0.901    34.533       0   \n",
       "                                 Nyamilu        -1.074    34.391       0   \n",
       "                                 Sumba          -0.686    34.596       0   \n",
       "                                 Uradi           0.980    34.543       0   \n",
       "                                 Uriri          -0.952    34.513       0   \n",
       "Malawi                           Chikwawa       -16.017   34.789       0   \n",
       "Mozambique                       Motinho        -10.851   40.594      40   \n",
       "                                 Palmeira       -25.255   32.873       0   \n",
       "                                 Palmeiras      -25.265   32.877       0   \n",
       "Nigeria                          Akaka Remo      6.950    3.717       41   \n",
       "Tanzania                         Minepa Village -8.273    36.681       0   \n",
       "Uganda                           Tororo          0.750    34.083       0   \n",
       "Zambia                           Katete         -14.183   31.867       0   \n",
       "\n",
       "year                                                                2016  \\\n",
       "country                          location       latitude longitude         \n",
       "Benin                            Kpome           6.383    2.200        0   \n",
       "Cameroon                         Mibellon        6.083    11.500       0   \n",
       "Central African Republic         Bangui          4.377    18.608      10   \n",
       "Democratic Republic of the Congo Kinshasa       -4.327    15.313       0   \n",
       "                                 Watsa Moke      3.094    29.548       0   \n",
       "Gabon                            Benguia        -1.633    13.492       0   \n",
       "Ghana                            Dimabi          9.420   -1.083        0   \n",
       "                                 Gbullung        9.488   -1.009        0   \n",
       "                                 Gupanarigu      9.497   -0.952        0   \n",
       "                                 Obuasi          5.933   -1.617        0   \n",
       "                                 Woribogu        8.869    0.321        0   \n",
       "Kenya                            Ahero          -0.174    34.920       0   \n",
       "                                 Awendo         -0.901    34.533       8   \n",
       "                                 Nyamilu        -1.074    34.391       1   \n",
       "                                 Sumba          -0.686    34.596       9   \n",
       "                                 Uradi           0.980    34.543      21   \n",
       "                                 Uriri          -0.952    34.513       5   \n",
       "Malawi                           Chikwawa       -16.017   34.789       0   \n",
       "Mozambique                       Motinho        -10.851   40.594       0   \n",
       "                                 Palmeira       -25.255   32.873      22   \n",
       "                                 Palmeiras      -25.265   32.877       0   \n",
       "Nigeria                          Akaka Remo      6.950    3.717        0   \n",
       "Tanzania                         Minepa Village -8.273    36.681       0   \n",
       "Uganda                           Tororo          0.750    34.083       0   \n",
       "Zambia                           Katete         -14.183   31.867      43   \n",
       "\n",
       "year                                                                2017  2018  \n",
       "country                          location       latitude longitude              \n",
       "Benin                            Kpome           6.383    2.200        0     0  \n",
       "Cameroon                         Mibellon        6.083    11.500       0     0  \n",
       "Central African Republic         Bangui          4.377    18.608       0     0  \n",
       "Democratic Republic of the Congo Kinshasa       -4.327    15.313       0     0  \n",
       "                                 Watsa Moke      3.094    29.548      43     0  \n",
       "Gabon                            Benguia        -1.633    13.492      40     0  \n",
       "Ghana                            Dimabi          9.420   -1.083        6     0  \n",
       "                                 Gbullung        9.488   -1.009       19     0  \n",
       "                                 Gupanarigu      9.497   -0.952       10     0  \n",
       "                                 Obuasi          5.933   -1.617        0     0  \n",
       "                                 Woribogu        8.869    0.321        1     0  \n",
       "Kenya                            Ahero          -0.174    34.920       0     0  \n",
       "                                 Awendo         -0.901    34.533       0     0  \n",
       "                                 Nyamilu        -1.074    34.391       0     0  \n",
       "                                 Sumba          -0.686    34.596       0     0  \n",
       "                                 Uradi           0.980    34.543       0     0  \n",
       "                                 Uriri          -0.952    34.513       0     0  \n",
       "Malawi                           Chikwawa       -16.017   34.789       0     0  \n",
       "Mozambique                       Motinho        -10.851   40.594       0     0  \n",
       "                                 Palmeira       -25.255   32.873       0     0  \n",
       "                                 Palmeiras      -25.265   32.877       0    76  \n",
       "Nigeria                          Akaka Remo      6.950    3.717        0     0  \n",
       "Tanzania                         Minepa Village -8.273    36.681      10     0  \n",
       "Uganda                           Tororo          0.750    34.083       0     0  \n",
       "Zambia                           Katete         -14.183   31.867       0     0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pivot_location_year = (\n",
    "    df_samples\n",
    "    .pivot_table(\n",
    "        index=[\"country\", \"location\", \"latitude\", \"longitude\"],\n",
    "        columns=[\"year\"],\n",
    "        values=\"sample_id\",\n",
    "        aggfunc=\"count\",\n",
    "        fill_value=0,\n",
    "    )\n",
    ")\n",
    "\n",
    "pivot_location_year"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be354317-4687-46b7-a273-e7c2bf20111d",
   "metadata": {},
   "source": [
    "Again, we can observe that most countries are only represented by a single location."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa597aeb-5034-4554-a149-4792e44e3609",
   "metadata": {},
   "source": [
    "Finally, we can represent the origin of these samples on a map for a more visual interpretation of their distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bf13a686-f0af-4e2a-a8a8-2480626b1dd1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "918729cd3e0e4ab4a0371524a7fc3d41",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map(center=[0, 20], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title', 'zoom_out_text‚Ä¶"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create a map\n",
    "m = ipyleaflet.Map(\n",
    "    basemap=ipyleaflet.basemaps.OpenStreetMap.Mapnik,\n",
    "    center=[0, 20],\n",
    "    zoom=3,\n",
    ")\n",
    "\n",
    "# add markers for sampling locations\n",
    "for row in pivot_location_year.reset_index().itertuples():\n",
    "    title = f\"{row.location}, {row.country} ({row.latitude:.3f}, {row.longitude:.3f})\"\n",
    "    for i, y in enumerate(pivot_location_year.columns):\n",
    "      r = list(pivot_location_year.loc[(row.country, row.location, row.latitude, row.longitude)])\n",
    "      if r[i] != 0:\n",
    "        title += \"\\n\" + str(r[i]) + \" samples in \" + str(y)\n",
    "    marker = ipyleaflet.Marker(\n",
    "        location=(row.latitude, row.longitude),\n",
    "        draggable=False,\n",
    "        title=title,\n",
    "    )\n",
    "    m.add_layer(marker)\n",
    "\n",
    "# add a scale bar\n",
    "m.add_control(ipyleaflet.ScaleControl(position=\"bottomleft\"))\n",
    "\n",
    "# display the map\n",
    "m"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4488bdc-68fa-4871-b1fb-a4aa24356fab",
   "metadata": {},
   "source": [
    "## Exploring the reference and finding genes of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b86acaa-e886-4645-9299-62ade6f600ea",
   "metadata": {},
   "source": [
    "We saw during the third module of the first workshop of the course on _An. gambiae_ ([Workshop 1 Module 2](https://anopheles-genomic-surveillance.github.io/workshop-1/module-3-genome.html)) how to access the genome annotations for that species. The same function exists for _An. funestus_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bfaba96e-b877-4175-8928-c6813ce17480",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Note</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>66492</td>\n",
       "      <td>69180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125765358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125765358, transcript varia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>66492</td>\n",
       "      <td>69180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125765358_t2</td>\n",
       "      <td>LOC125765358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125765358, transcript varia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>66492</td>\n",
       "      <td>69180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125765358_t1</td>\n",
       "      <td>LOC125765358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125765358, transcript varia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>66492</td>\n",
       "      <td>66764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon_LOC125765358_t2-E1</td>\n",
       "      <td>LOC125765358_t2,LOC125765358_t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>66923</td>\n",
       "      <td>69180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125765358_t3</td>\n",
       "      <td>LOC125765358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125765358, transcript varia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347419</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>22261674</td>\n",
       "      <td>22261945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon_LOC125771994_t1-E5</td>\n",
       "      <td>LOC125771994_t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347420</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>22262762</td>\n",
       "      <td>22262829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon_LOC125771994_t1-E4</td>\n",
       "      <td>LOC125771994_t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347421</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>22263509</td>\n",
       "      <td>22263916</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon_LOC125771994_t1-E3</td>\n",
       "      <td>LOC125771994_t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347422</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>22263985</td>\n",
       "      <td>22264052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon_LOC125771994_t1-E2</td>\n",
       "      <td>LOC125771994_t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347423</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>22264121</td>\n",
       "      <td>22264183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>exon_LOC125771994_t1-E1</td>\n",
       "      <td>LOC125771994_t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347424 rows √ó 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       contig     source        type     start       end  score strand  phase  \\\n",
       "0         2RL  VEuPathDB  ncRNA_gene     66492     69180    NaN      +    NaN   \n",
       "1         2RL  VEuPathDB     lnc_RNA     66492     69180    NaN      +    NaN   \n",
       "2         2RL  VEuPathDB     lnc_RNA     66492     69180    NaN      +    NaN   \n",
       "3         2RL  VEuPathDB        exon     66492     66764    NaN      +    NaN   \n",
       "4         2RL  VEuPathDB     lnc_RNA     66923     69180    NaN      +    NaN   \n",
       "...       ...        ...         ...       ...       ...    ...    ...    ...   \n",
       "347419      X  VEuPathDB        exon  22261674  22261945    NaN      -    NaN   \n",
       "347420      X  VEuPathDB        exon  22262762  22262829    NaN      -    NaN   \n",
       "347421      X  VEuPathDB        exon  22263509  22263916    NaN      -    NaN   \n",
       "347422      X  VEuPathDB        exon  22263985  22264052    NaN      -    NaN   \n",
       "347423      X  VEuPathDB        exon  22264121  22264183    NaN      -    NaN   \n",
       "\n",
       "                             ID                           Parent Note  \\\n",
       "0                  LOC125765358                              NaN  NaN   \n",
       "1               LOC125765358_t2                     LOC125765358  NaN   \n",
       "2               LOC125765358_t1                     LOC125765358  NaN   \n",
       "3       exon_LOC125765358_t2-E1  LOC125765358_t2,LOC125765358_t1  NaN   \n",
       "4               LOC125765358_t3                     LOC125765358  NaN   \n",
       "...                         ...                              ...  ...   \n",
       "347419  exon_LOC125771994_t1-E5                  LOC125771994_t1  NaN   \n",
       "347420  exon_LOC125771994_t1-E4                  LOC125771994_t1  NaN   \n",
       "347421  exon_LOC125771994_t1-E3                  LOC125771994_t1  NaN   \n",
       "347422  exon_LOC125771994_t1-E2                  LOC125771994_t1  NaN   \n",
       "347423  exon_LOC125771994_t1-E1                  LOC125771994_t1  NaN   \n",
       "\n",
       "                                              description  \n",
       "0       uncharacterized LOC125765358, transcript varia...  \n",
       "1       uncharacterized LOC125765358, transcript varia...  \n",
       "2       uncharacterized LOC125765358, transcript varia...  \n",
       "3                                                     NaN  \n",
       "4       uncharacterized LOC125765358, transcript varia...  \n",
       "...                                                   ...  \n",
       "347419                                                NaN  \n",
       "347420                                                NaN  \n",
       "347421                                                NaN  \n",
       "347422                                                NaN  \n",
       "347423                                                NaN  \n",
       "\n",
       "[347424 rows x 12 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f2c90ef-fbc6-428c-9045-dc40cfd464b6",
   "metadata": {},
   "source": [
    "The first big difference between the reference for _An. funestus_ and _An. gambiae_ is that contigs are whole chromosomes (e.g., \"2RL\") instead of chromosome arms (e.g., \"2R\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b42d21d7-f992-4c3e-a9a7-2f30259703cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "contig\n",
       "2RL    170850\n",
       "3RL    141790\n",
       "X       34784\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features().groupby(\"contig\").size()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06d6645d-b637-4f78-9d3a-8e6c9f26bc06",
   "metadata": {},
   "source": [
    "We can group the annotations by \"type\" to see what the annotations contain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "489ee49a-3ea8-4bf4-9a29-a9cb55a37885",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "type\n",
       "CDS                       161215\n",
       "exon                       78083\n",
       "five_prime_UTR             42054\n",
       "lnc_RNA                      942\n",
       "mRNA                       25010\n",
       "ncRNA                        609\n",
       "ncRNA_gene                  1041\n",
       "protein_coding_gene        11962\n",
       "pseudogene                    42\n",
       "pseudogenic_transcript        42\n",
       "rRNA                         101\n",
       "snRNA                         27\n",
       "snoRNA                         9\n",
       "tRNA                         316\n",
       "three_prime_UTR            25971\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features().groupby(\"type\").size()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4142b686-67db-4c21-8ab4-6520913686e5",
   "metadata": {},
   "source": [
    "Many different types of annotations are provided but we will only be interested in \"protein_coding_gene\"s for this module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c5c9703c-df9a-4991-b43a-e9af4dbcf9fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Note</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>81055</td>\n",
       "      <td>82221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125766088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60S ribosomal protein L18 [Source:UniProtKB/Tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>82584</td>\n",
       "      <td>84996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125766087</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Folliculin [Source:UniProtKB/TrEMBL;Acc:A0A182...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>141143</td>\n",
       "      <td>146973</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125766196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125766196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>159192</td>\n",
       "      <td>185030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125765747</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lachesin-like, transcript variant X5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>227643</td>\n",
       "      <td>228822</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125760444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>actin-related protein 2/3 complex subunit 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347372</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>22057024</td>\n",
       "      <td>22057530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125768238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcriptional regulatory protein AlgP-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347378</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>22078120</td>\n",
       "      <td>22083556</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125769511</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125769511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347386</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>22089764</td>\n",
       "      <td>22093137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125768244</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125768244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347394</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>22192285</td>\n",
       "      <td>22193524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125768250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein lag-3-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347402</th>\n",
       "      <td>X</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>22206931</td>\n",
       "      <td>22212078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125769281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>uncharacterized LOC125769281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11962 rows √ó 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       contig     source                 type     start       end  score  \\\n",
       "16        2RL  VEuPathDB  protein_coding_gene     81055     82221    NaN   \n",
       "26        2RL  VEuPathDB  protein_coding_gene     82584     84996    NaN   \n",
       "40        2RL  VEuPathDB  protein_coding_gene    141143    146973    NaN   \n",
       "50        2RL  VEuPathDB  protein_coding_gene    159192    185030    NaN   \n",
       "142       2RL  VEuPathDB  protein_coding_gene    227643    228822    NaN   \n",
       "...       ...        ...                  ...       ...       ...    ...   \n",
       "347372      X  VEuPathDB  protein_coding_gene  22057024  22057530    NaN   \n",
       "347378      X  VEuPathDB  protein_coding_gene  22078120  22083556    NaN   \n",
       "347386      X  VEuPathDB  protein_coding_gene  22089764  22093137    NaN   \n",
       "347394      X  VEuPathDB  protein_coding_gene  22192285  22193524    NaN   \n",
       "347402      X  VEuPathDB  protein_coding_gene  22206931  22212078    NaN   \n",
       "\n",
       "       strand  phase            ID Parent Note  \\\n",
       "16          +    NaN  LOC125766088    NaN  NaN   \n",
       "26          -    NaN  LOC125766087    NaN  NaN   \n",
       "40          +    NaN  LOC125766196    NaN  NaN   \n",
       "50          -    NaN  LOC125765747    NaN  NaN   \n",
       "142         +    NaN  LOC125760444    NaN  NaN   \n",
       "...       ...    ...           ...    ...  ...   \n",
       "347372      -    NaN  LOC125768238    NaN  NaN   \n",
       "347378      +    NaN  LOC125769511    NaN  NaN   \n",
       "347386      -    NaN  LOC125768244    NaN  NaN   \n",
       "347394      +    NaN  LOC125768250    NaN  NaN   \n",
       "347402      +    NaN  LOC125769281    NaN  NaN   \n",
       "\n",
       "                                              description  \n",
       "16      60S ribosomal protein L18 [Source:UniProtKB/Tr...  \n",
       "26      Folliculin [Source:UniProtKB/TrEMBL;Acc:A0A182...  \n",
       "40                           uncharacterized LOC125766196  \n",
       "50                   lachesin-like, transcript variant X5  \n",
       "142           actin-related protein 2/3 complex subunit 4  \n",
       "...                                                   ...  \n",
       "347372       transcriptional regulatory protein AlgP-like  \n",
       "347378                       uncharacterized LOC125769511  \n",
       "347386                       uncharacterized LOC125768244  \n",
       "347394                                 protein lag-3-like  \n",
       "347402                       uncharacterized LOC125769281  \n",
       "\n",
       "[11962 rows x 12 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features().query(\"type == 'protein_coding_gene'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ea3ec4-e36c-446b-a17a-bcb0bf7669cc",
   "metadata": {},
   "source": [
    "We can see that the naming convention (\"LOC?????????\") is different from the one used for _An. gambiae_ (\"AGAP??????\")."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3317e384-194f-4816-bf66-79369e223ecb",
   "metadata": {},
   "source": [
    "In order to locate genes of interest, several methods can be used. The easiest is when the ID is already known. For instance, in supplementary Table 3 of [15], the ID for GSTE2 is given as \"LOC125763977\". We can thus look at this gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "22819c04-aa2c-431d-9951-70a079e94af4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Note</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133312</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>76406705</td>\n",
       "      <td>76407655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125763977</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>glutathione S-transferase 1-like, transcript v...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       contig     source                 type     start       end  score  \\\n",
       "133312    2RL  VEuPathDB  protein_coding_gene  76406705  76407655    NaN   \n",
       "\n",
       "       strand  phase            ID Parent Note  \\\n",
       "133312      +    NaN  LOC125763977    NaN  NaN   \n",
       "\n",
       "                                              description  \n",
       "133312  glutathione S-transferase 1-like, transcript v...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features().query(\"ID == 'LOC125763977'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05963ea6-ccc4-43ed-b9da-0f6857f6ae0c",
   "metadata": {},
   "source": [
    "Another fairly simple case is when we know what the description of the gene is. For instance, if we are interested in ACE1, its description is \"Acetylcholinesterase [Source:UniProtKB/TrEMBL;Acc:A0A182RZ85]\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2ac12ad2-f1fe-4966-85bf-5a19fe68156b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Note</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38520</th>\n",
       "      <td>2RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>19417355</td>\n",
       "      <td>19490903</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125763056</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acetylcholinesterase [Source:UniProtKB/TrEMBL;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      contig     source                 type     start       end  score  \\\n",
       "38520    2RL  VEuPathDB  protein_coding_gene  19417355  19490903    NaN   \n",
       "\n",
       "      strand  phase            ID Parent Note  \\\n",
       "38520      +    NaN  LOC125763056    NaN  NaN   \n",
       "\n",
       "                                             description  \n",
       "38520  Acetylcholinesterase [Source:UniProtKB/TrEMBL;...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features().query(\"description == 'Acetylcholinesterase [Source:UniProtKB/TrEMBL;Acc:A0A182RZ85]'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53c0fe73-bae3-4be7-86c6-4ab17e2c3d1c",
   "metadata": {},
   "source": [
    "The \"[Source:UniProtKB/TrEMBL;Acc:A0A182RZ85]\" part makes it hard to remember and is thus problematic. It is possible, though not really simple, to dispense with it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "45bece94-e83d-4352-8ef8-24eb6afabed1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "contig                                                       2RL\n",
       "source                                                 VEuPathDB\n",
       "type                                         protein_coding_gene\n",
       "start                                                   19417355\n",
       "end                                                     19490903\n",
       "score                                                        NaN\n",
       "strand                                                         +\n",
       "phase                                                        NaN\n",
       "ID                                                  LOC125763056\n",
       "Parent                                                       NaN\n",
       "Note                                                         NaN\n",
       "description    Acetylcholinesterase [Source:UniProtKB/TrEMBL;...\n",
       "Name: 38520, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "correct_id  = [i for i, d in enumerate(af1.genome_features()['description']) if type(d) == str if d.split(\" \")[0] == 'Acetylcholinesterase']\n",
    "af1.genome_features().loc[correct_id[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c579d5bf-3fcf-43a1-a2fa-cdea19463914",
   "metadata": {},
   "source": [
    "Another way to find genes of interest is by looking at specific regions of the genome, for instance ones that are around signals of selection. For instance, if we look at the region \"X:8445000-8455000\", we find Cyp9k1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e30ec8d9-4030-4c30-9e75-fdc76aaefa24",
   "metadata": {},
   "outputs": [],
   "source": [
    "af1.genome_features(region=\"X:8445000-8455000\").query(\"type == 'protein_coding_gene'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4ee750c-4aa5-46cb-b06e-31a9f284e4b2",
   "metadata": {},
   "source": [
    "The most common way to find a gene, however, is to use [Vectorbase](https://vectorbase.org/vectorbase/app) and look for orthologs for known genes in other species. For instance, if one looks at the orthologs of AGAP004707 (VGSC in _An. gambiae_), one can find several proteins of the gene AFUN2_008728 of AfunGA1 which can be found in the region \"3RL:44105643-44156624\". If we look at this region, we find the gene VGSC for _An. funestus_."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b0b73dfe-0749-4e73-a099-537e614ccc42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Note</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3RL</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>44105643</td>\n",
       "      <td>44156624</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC125769886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sodium channel protein para, transcript varian...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contig     source                 type     start       end  score strand  \\\n",
       "0    3RL  VEuPathDB  protein_coding_gene  44105643  44156624    NaN      -   \n",
       "\n",
       "   phase            ID Parent Note  \\\n",
       "0    NaN  LOC125769886    NaN  NaN   \n",
       "\n",
       "                                         description  \n",
       "0  sodium channel protein para, transcript varian...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af1.genome_features(region=\"3RL:44105643-44156624\").query(\"type == 'protein_coding_gene'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78978e18-1f0d-4d94-915b-1750faccc6a5",
   "metadata": {},
   "source": [
    "1. Sinka, M.E., Bangs, M.J., Manguin, S. et al. A global map of dominant malaria vectors. Parasites Vectors 5, 69 (2012). https://doi.org/10.1186/1756-3305-5-69\n",
    "2. Msugupakulya, B.J., Urio, N.H., Jumanne, M. et al. Changes in contributions of different Anopheles vector species to malaria transmission in east and southern Africa from 2000 to 2022. Parasites Vectors 16, 408 (2023). https://doi.org/10.1186/s13071-023-06019-1\n",
    "3. McCann RS, Ochomo E, Bayoh MN, Vulule JM, Hamel MJ, Gimnig JE, Hawley WA, Walker ED. Reemergence of Anopheles funestus as a vector of Plasmodium falciparum in western Kenya after long-term implementation of insecticide-treated bed nets. Am J Trop Med Hyg. 2014 Apr;90(4):597-604. doi: 10.4269/ajtmh.13-0614. Epub 2014 Jan 27. PMID: 24470562; PMCID: PMC3973499.\n",
    "4. Ntabaliba W, Vavassori L, Stica C, Makungwa N, Odufuwa OG, Swai JK, Lekundayo R, Moore S. Life expectancy of Anopheles funestus is double that of Anopheles arabiensis in southeast Tanzania based on mark-release-recapture method. Sci Rep. 2023 Sep 22;13(1):15775. doi: 10.1038/s41598-023-42761-3. PMID: 37737323; PMCID: PMC10516982.\n",
    "5. Ogola EO, Fillinger U, Ondiba IM, Villinger J, Masiga DK, Torto B, Tchouassi DP. Insights into malaria transmission among Anopheles funestus mosquitoes, Kenya. Parasit Vectors. 2018 Nov 6;11(1):577. doi: 10.1186/s13071-018-3171-3. PMID: 30400976; PMCID: PMC6219006.\n",
    "6. Djouaka R, Akoton R, Tchigossou GM, Atoyebi SM, Irving H, Kusimo MO, Djegbe I, Riveron JM, Tossou E, Yessoufou A, Wondji CS. Mapping the distribution of Anopheles funestus across Benin highlights a sharp contrast of susceptibility to insecticides and infection rate to Plasmodium between southern and northern populations. Wellcome Open Res. 2016 Dec 14;1:28. doi: 10.12688/wellcomeopenres.10213.2. PMID: 28191507; PMCID: PMC5300096.\n",
    "7. Dia I., Wamdaogo M., Ayala D., ‚ÄúAdvances and Perspectives in the Study of the Malaria Mosquito Anopheles funestus‚Äù in Anopheles Mosquitoes - New Insights into Malaria Vectors, Manguin S., Ed. (InTech, 2013). \n",
    "8. Moiroux N, Gomez MB, Pennetier C, Elanga E, Dj√®nontin A, Chandre F, Dj√®gb√© I, Guis H, Corbel V. Changes in Anopheles funestus biting behavior following universal coverage of long-lasting insecticidal nets in Benin. J Infect Dis. 2012 Nov 15;206(10):1622-9. doi: 10.1093/infdis/jis565. Epub 2012 Sep 10. PMID: 22966127.\n",
    "9. Sougoufara S, Di√©dhiou SM, Doucour√© S, Diagne N, Semb√®ne PM, Harry M, Trape JF, Sokhna C, Ndiath MO. Biting by Anopheles funestus in broad daylight after use of long-lasting insecticidal nets: a new challenge to malaria elimination. Malar J. 2014 Mar 28;13:125. doi: 10.1186/1475-2875-13-125. PMID: 24678587; PMCID: PMC3973838.\n",
    "10. Omondi S, Kosgei J, Musula G, Muchoki M, Abong'o B, Agumba S, Ogwang C, McDermott DP, Donnelly MJ, Staedke SG, Schultz J, Gutman JR, Gimnig JE, Ochomo E. Late morning biting behaviour of Anopheles funestus is a risk factor for transmission in schools in Siaya, western Kenya. Malar J. 2023 Nov 30;22(1):366. doi: 10.1186/s12936-023-04806-w. PMID: 38037026; PMCID: PMC10691009.\n",
    "11. Nambunga IH, Ngowo HS, Mapua SA, Hape EE, Msugupakulya BJ, Msaky DS, Mhumbira NT, Mchwembo KR, Tamayamali GZ, Mlembe SV, Njalambaha RM, Lwetoijera DW, Finda MF, Govella NJ, Matoke-Muhia D, Kaindoa EW, Okumu FO. Aquatic habitats of the malaria vector Anopheles funestus in rural south-eastern Tanzania. Malar J. 2020 Jun 23;19(1):219. doi: 10.1186/s12936-020-03295-5. PMID: 32576200; PMCID: PMC7310514.\n",
    "12. Jones, C.M., Lee, Y., Kitchen, A. et al. Complete Anopheles funestus mitogenomes reveal an ancient history of mitochondrial lineages and their distribution in southern and central Africa. Sci Rep 8, 9054 (2018). https://doi.org/10.1038/s41598-018-27092-y\n",
    "13. The origin of our species. National History Museum, London, UK. https://www.nhm.ac.uk/discover/the-origin-of-our-species.html\n",
    "14. Ayala D, Akone-Ella O, Kengne P, Johnson H, Heaton H, Collins J, Krasheninnikova K, Pelan S, Pointon DL, Sims Y, Torrance J, Tracey A, Uliano-Silva M, von Wyschetzki K, Wood J; DNA Pipelines collective; McCarthy S, Neafsey D, Makunin A, Lawniczak M. The genome sequence of the malaria mosquito, Anopheles funestus, Giles, 1900. Wellcome Open Res. 2023 Mar 27;7:287. doi: 10.12688/wellcomeopenres.18445.2. PMID: 36874567; PMCID: PMC9975407."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "877d2123-0dbc-4306-947d-4968768bb153",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "developer-developer-training-nb-maintenance-mgen-15.2.0",
   "language": "python",
   "name": "conda-env-developer-developer-training-nb-maintenance-mgen-15.2.0-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
