
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Module 4 - Analysing SNPs in the Vgsc gene &#8212; Training course in data analysis for genomic surveillance of African malaria vectors</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet">
  <link href="../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/myfile.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Module 3 - Navigating the Anopheles gambiae reference genome" href="module-3-genome.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Training course in data analysis for genomic surveillance of African malaria vectors</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../home.html">
   Home
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Workshop 1
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="about.html">
   Workshop 1 - Pyrethroid target-site resistance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="module-1-colab.html">
   Module 1 - Interactive cloud computing with Google Colaboratory
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="module-2-sample-metadata.html">
   Module 2 - Accessing and exploring
   <em>
    Anopheles
   </em>
   genomic data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="module-3-genome.html">
   Module 3 - Navigating the
   <em>
    Anopheles gambiae
   </em>
   reference genome
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Module 4 - Analysing SNPs in the
   <em>
    Vgsc
   </em>
   gene
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/workshop-1/module-4-vgsc-snps.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/issues/new?title=Issue%20on%20page%20%2Fworkshop-1/module-4-vgsc-snps.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/docs/workshop-1/module-4-vgsc-snps.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning objectives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#lecture">
   Lecture
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#english">
     English
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#francais">
     Français
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#discovering-snps-in-the-vgsc-gene-and-computing-snp-allele-frequencies">
   Discovering SNPs in the
   <em>
    Vgsc
   </em>
   gene and computing SNP allele frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-samples-into-cohorts">
   Grouping samples into cohorts
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#single-nucleotide-polymorphisms-snps">
   Single nucleotide polymorphisms (SNPs)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-effects-some-snps-may-be-more-interesting-than-others">
   SNP effects - some SNPs may be more interesting than others
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-allele-frequencies">
   SNP allele frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#visualising-snp-allele-frequencies">
   Visualising SNP allele frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#amino-acid-substitution-frequencies">
   Amino acid substitution frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#well-done">
   Well done!
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#practical-exercises">
   Practical exercises
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     English
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Français
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Module 4 - Analysing SNPs in the Vgsc gene</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning objectives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#lecture">
   Lecture
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#english">
     English
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#francais">
     Français
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#discovering-snps-in-the-vgsc-gene-and-computing-snp-allele-frequencies">
   Discovering SNPs in the
   <em>
    Vgsc
   </em>
   gene and computing SNP allele frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-samples-into-cohorts">
   Grouping samples into cohorts
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#single-nucleotide-polymorphisms-snps">
   Single nucleotide polymorphisms (SNPs)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-effects-some-snps-may-be-more-interesting-than-others">
   SNP effects - some SNPs may be more interesting than others
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snp-allele-frequencies">
   SNP allele frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#visualising-snp-allele-frequencies">
   Visualising SNP allele frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#amino-acid-substitution-frequencies">
   Amino acid substitution frequencies
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#well-done">
   Well done!
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#practical-exercises">
   Practical exercises
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     English
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Français
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <p><em><strong>Training course in data analysis for genomic surveillance of African malaria vectors - Workshop 1</strong></em></p>
<hr class="docutils" />
<div class="tex2jax_ignore mathjax_ignore section" id="module-4-analysing-snps-in-the-vgsc-gene">
<h1>Module 4 - Analysing SNPs in the <em>Vgsc</em> gene<a class="headerlink" href="#module-4-analysing-snps-in-the-vgsc-gene" title="Permalink to this headline">¶</a></h1>
<p><strong>Theme: Analysis</strong></p>
<p>In this module we’re going to perform an analysis to discover single nucleotide polymorphisms (SNPs) in the voltage-gated sodium channel gene (<em>Vgsc</em>), which encodes the binding target for pyrethroid insecticides.</p>
<div class="section" id="learning-objectives">
<h2>Learning objectives<a class="headerlink" href="#learning-objectives" title="Permalink to this headline">¶</a></h2>
<p>At the end of this module you will be able to:</p>
<ul class="simple">
<li><p>Discover mutations (SNPs) that could potentially cause pyrethroid target-site resistance.</p></li>
<li><p>Compute SNP allele frequencies, i.e., how common are they in different mosquito cohorts?</p></li>
<li><p>Perform analyses to compare SNP allele frequences between mosquitoes from different species, geographical locations and dates of collection.</p></li>
</ul>
</div>
<div class="section" id="lecture">
<h2>Lecture<a class="headerlink" href="#lecture" title="Permalink to this headline">¶</a></h2>
<div class="section" id="english">
<h3>English<a class="headerlink" href="#english" title="Permalink to this headline">¶</a></h3>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/m3R5SuNgKmw" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></div></div>
</div>
</div>
<div class="section" id="francais">
<h3>Français<a class="headerlink" href="#francais" title="Permalink to this headline">¶</a></h3>
<p>Coming soon.</p>
</div>
</div>
<div class="section" id="discovering-snps-in-the-vgsc-gene-and-computing-snp-allele-frequencies">
<h2>Discovering SNPs in the <em>Vgsc</em> gene and computing SNP allele frequencies<a class="headerlink" href="#discovering-snps-in-the-vgsc-gene-and-computing-snp-allele-frequencies" title="Permalink to this headline">¶</a></h2>
<p>First, let’s set up the <a class="reference external" href="https://github.com/malariagen/malariagen-data-python"><code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code></a> package to access MalariaGEN data in the cloud.</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip install -q malariagen_data
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">malariagen_data</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Ag3</span><span class="p">()</span>
<span class="n">ag3</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<style type="text/css">
    table.malariagen-ag3 th, table.malariagen-ag3 td {
        text-align: left
    }
</style>
<table class="malariagen-ag3">
    <thead>
        <tr>
            <th colspan=2>MalariaGEN Ag3 data resource API</th>
        </tr>
    </thead>
    <tbody>
        <tr><th>Storage URL</th><td>gs://vo_agam_release/</td></tr>
        <tr><th>Releases available</th><td>3.0</td></tr>
        <tr><th>Cohorts analysis</th><td>20211101</td></tr>
        <tr><th>Species analysis</th><td>aim_20200422</td></tr>
        <tr><th>Site filters analysis</th><td>dt_20200416</td></tr>
    </tbody>
</table>
<p>Please note that data are subject to terms of use,
for more information see <a href="https://www.malariagen.net/data">
the MalariaGEN website</a> or contact data@malariagen.net.</p>
<p>See also the <a href="https://malariagen.github.io/vector-data/ag3/api.html">Ag3 API docs</a>.</p>
</div></div>
</div>
<p>To discover SNPs and compute allele frequencies, we’re going to use the <a class="reference external" href="https://malariagen.github.io/vector-data/ag3/api.html#snp-allele-frequencies"><code class="docutils literal notranslate"><span class="pre">snp_allele_frequencies()</span></code> function</a>. Let’s have a look at the documentation for this function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ag3.snp_allele_frequencies<span class="o">?</span>
</pre></div>
</div>
</div>
</div>
<p>To discover SNPs in the <em>Vgsc</em> gene, we need to define some parameters.</p>
<p>First, we need to decide which gene transcript to use when determining what SNP effects will be. Here we’ll use the transcript with identifier “AGAP004707-RD”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">transcript</span> <span class="o">=</span> <span class="s2">&quot;AGAP004707-RD&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>Next, to compute allele frequencies, we need to decide how our mosquitoes will be grouped into cohorts. There are different ways you can do this, for this analysis we’ll group spatially by level 1 administrative divisions within countries, and temporally by year.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cohorts</span> <span class="o">=</span> <span class="s2">&quot;admin1_year&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we need to choose which samples to analyse. There are a number of different sample sets in the Ag3.0 data resource that we could use for this analysis. Let’s check what’s available.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">sample_sets</span><span class="p">(</span><span class="n">release</span><span class="o">=</span><span class="s2">&quot;3.0&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-65869631-83eb-44aa-be37-ddfd60cf32f6">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_set</th>
      <th>sample_count</th>
      <th>release</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AG1000G-AO</td>
      <td>81</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AG1000G-BF-A</td>
      <td>181</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AG1000G-BF-B</td>
      <td>102</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AG1000G-BF-C</td>
      <td>13</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AG1000G-CD</td>
      <td>76</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AG1000G-CF</td>
      <td>73</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>AG1000G-CI</td>
      <td>80</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>AG1000G-CM-A</td>
      <td>303</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>AG1000G-CM-B</td>
      <td>97</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>AG1000G-CM-C</td>
      <td>44</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>AG1000G-FR</td>
      <td>23</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>11</th>
      <td>AG1000G-GA-A</td>
      <td>69</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>12</th>
      <td>AG1000G-GH</td>
      <td>100</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>13</th>
      <td>AG1000G-GM-A</td>
      <td>74</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>14</th>
      <td>AG1000G-GM-B</td>
      <td>31</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>15</th>
      <td>AG1000G-GM-C</td>
      <td>174</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>16</th>
      <td>AG1000G-GN-A</td>
      <td>45</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>17</th>
      <td>AG1000G-GN-B</td>
      <td>185</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>18</th>
      <td>AG1000G-GQ</td>
      <td>10</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>19</th>
      <td>AG1000G-GW</td>
      <td>101</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>20</th>
      <td>AG1000G-KE</td>
      <td>86</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>21</th>
      <td>AG1000G-ML-A</td>
      <td>60</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>22</th>
      <td>AG1000G-ML-B</td>
      <td>71</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>23</th>
      <td>AG1000G-MW</td>
      <td>41</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>24</th>
      <td>AG1000G-MZ</td>
      <td>74</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>25</th>
      <td>AG1000G-TZ</td>
      <td>300</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>26</th>
      <td>AG1000G-UG</td>
      <td>290</td>
      <td>3.0</td>
    </tr>
    <tr>
      <th>27</th>
      <td>AG1000G-X</td>
      <td>297</td>
      <td>3.0</td>
    </tr>
  </tbody>
</table>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-65869631-83eb-44aa-be37-ddfd60cf32f6')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-65869631-83eb-44aa-be37-ddfd60cf32f6 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-65869631-83eb-44aa-be37-ddfd60cf32f6');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>To keep things simple, for this module we’ll focus on mosquitoes from Burkina Faso. There are three sample sets in the Ag3.0 resource providing data on mosquitoes from Burkina Faso.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample_sets</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-C&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>OK, now we’re ready to run the analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_allele_freqs_df</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_allele_frequencies</span><span class="p">(</span>
    <span class="n">transcript</span><span class="o">=</span><span class="n">transcript</span><span class="p">,</span> 
    <span class="n">cohorts</span><span class="o">=</span><span class="n">cohorts</span><span class="p">,</span> 
    <span class="n">sample_sets</span><span class="o">=</span><span class="n">sample_sets</span><span class="p">,</span> 
    <span class="n">drop_invariant</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">snp_allele_freqs_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-1ea0aa58-0390-4c26-9dcd-b18882b09019">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>pass_gamb_colu_arab</th>
      <th>pass_gamb_colu</th>
      <th>pass_arab</th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>frq_BF-07_gamb_2004</th>
      <th>max_af</th>
      <th>effect</th>
      <th>impact</th>
      <th>ref_codon</th>
      <th>alt_codon</th>
      <th>aa_pos</th>
      <th>ref_aa</th>
      <th>alt_aa</th>
    </tr>
    <tr>
      <th>contig</th>
      <th>position</th>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="11" valign="top">2L</th>
      <th rowspan="3" valign="top">2358158</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>M1L</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>START_LOST</td>
      <td>HIGH</td>
      <td>Atg</td>
      <td>Ctg</td>
      <td>1.0</td>
      <td>M</td>
      <td>L</td>
    </tr>
    <tr>
      <th>T</th>
      <th>M1L</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>START_LOST</td>
      <td>HIGH</td>
      <td>Atg</td>
      <td>Ttg</td>
      <td>1.0</td>
      <td>M</td>
      <td>L</td>
    </tr>
    <tr>
      <th>G</th>
      <th>M1V</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>START_LOST</td>
      <td>HIGH</td>
      <td>Atg</td>
      <td>Gtg</td>
      <td>1.0</td>
      <td>M</td>
      <td>V</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2358159</th>
      <th rowspan="2" valign="top">T</th>
      <th>A</th>
      <th>M1K</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>aTg</td>
      <td>aAg</td>
      <td>1.0</td>
      <td>M</td>
      <td>K</td>
    </tr>
    <tr>
      <th>C</th>
      <th>M1T</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>aTg</td>
      <td>aCg</td>
      <td>1.0</td>
      <td>M</td>
      <td>T</td>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2431616</th>
      <th rowspan="2" valign="top">G</th>
      <th>C</th>
      <th>*2119S</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>STOP_LOST</td>
      <td>HIGH</td>
      <td>tGa</td>
      <td>tCa</td>
      <td>2119.0</td>
      <td>*</td>
      <td>S</td>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119L</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>STOP_LOST</td>
      <td>HIGH</td>
      <td>tGa</td>
      <td>tTa</td>
      <td>2119.0</td>
      <td>*</td>
      <td>L</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">2431617</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>*2119C</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>STOP_LOST</td>
      <td>HIGH</td>
      <td>tgA</td>
      <td>tgC</td>
      <td>2119.0</td>
      <td>*</td>
      <td>C</td>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119C</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>STOP_LOST</td>
      <td>HIGH</td>
      <td>tgA</td>
      <td>tgT</td>
      <td>2119.0</td>
      <td>*</td>
      <td>C</td>
    </tr>
    <tr>
      <th>G</th>
      <th>*2119W</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>STOP_LOST</td>
      <td>HIGH</td>
      <td>tgA</td>
      <td>tgG</td>
      <td>2119.0</td>
      <td>*</td>
      <td>W</td>
    </tr>
  </tbody>
</table>
<p>209001 rows × 16 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-1ea0aa58-0390-4c26-9dcd-b18882b09019')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-1ea0aa58-0390-4c26-9dcd-b18882b09019 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-1ea0aa58-0390-4c26-9dcd-b18882b09019');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>The output from this function is a pandas DataFrame, where each row provides information about a SNP in the <em>Vgsc</em> gene.</p>
<p>Before we go further, to improve our understanding of what these data, let’s look at some background.</p>
</div>
<div class="section" id="grouping-samples-into-cohorts">
<h2>Grouping samples into cohorts<a class="headerlink" href="#grouping-samples-into-cohorts" title="Permalink to this headline">¶</a></h2>
<p>The MalariaGEN Ag3.0 data resource contain mosquito samples collected across large spatial and temporal scales, and from different mosquito species. When we want to run population genetic analyses on datasets like these, the data must be divided into biologically relevant <strong>cohorts</strong>, where a cohort is simply a group of samples we want to analyse together.</p>
<p>To help define cohorts and analyse these data, we have added some metadata for each sample about its time and place of collection and its species:</p>
<ul class="simple">
<li><p><strong>Spatially</strong> - For most analyses we use administrative divisions to group the samples into cohorts. These give two levels of spatial resolution, where admin level 1 divides each country into a few large regions, while admin level 2 provides finer scale divisions such as provinces.</p></li>
<li><p><strong>Temporally</strong> - For each sample we provide the year and month of collection. Depending on your analysis, you can choose to group samples by year or by year and by month, although note that for some samples the collection month is missing.</p></li>
<li><p><strong>Taxonomically</strong> - Ag3.0 contains samples from different species in the <em>Anopheles gambiae</em> complex. To help with grouping by taxon, we have included a “taxon” field in the sample metadata.</p></li>
</ul>
<p>Using these three dimensions, we have pre-defined four <strong>cohort sets</strong>, each of which groups samples into cohorts at different levels of spatio-temporal resolution. Within all cohort sets, samples are further subdivided by taxon.</p>
<ul class="simple">
<li><p><strong>admin1_year</strong> - Cohorts obtained by grouping samples by admin level 1, collection year and taxon.</p></li>
<li><p><strong>admin1_month</strong> - Cohorts obtained by grouping samples by admin level 1, collection year and month, and taxon.</p></li>
<li><p><strong>admin2_year</strong> - Cohorts obtained by grouping samples by admin level 2, collection year and taxon.</p></li>
<li><p><strong>admin2_month</strong> - Cohorts obtained by grouping samples by admin level 2, collection year and month, and taxon.</p></li>
</ul>
<p>Remember above we chose to use the “admin1_year” cohorts for our <em>Vgsc</em> analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cohorts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;admin1_year&#39;
</pre></div>
</div>
</div>
</div>
<p>Let’s now use this to understand the frequency columns in the SNP allele frequencies DataFrame. Here are the frequency column names:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">frequency_columns</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">snp_allele_freqs_df</span><span class="o">.</span><span class="n">columns</span> 
    <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;frq_&quot;</span><span class="p">)</span>
<span class="p">]</span>
<span class="n">frequency_columns</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;frq_BF-09_gamb_2012&#39;,
 &#39;frq_BF-09_colu_2012&#39;,
 &#39;frq_BF-09_colu_2014&#39;,
 &#39;frq_BF-09_gamb_2014&#39;,
 &#39;frq_BF-07_gamb_2004&#39;]
</pre></div>
</div>
</div>
</div>
<p>Here “frq_” is used to mean that these columns contain frequencies.</p>
<p>The second part of the column name is either “BF-09” or “BF-07”. These are <a class="reference external" href="https://en.wikipedia.org/wiki/ISO_3166-2">standard identifiers</a> that refer to level 1 administrative divisions within countries, often called regions. “BF-07” is the Centre-Sud region within Burkina Faso, and “BF-09” is the Haut-Bassins region.</p>
<p>The third part refers to the species, and in this example is either “gamb” meaning <em>Anopheles gambiae</em> or “colu” meaning <em>Anopheles coluzzii</em>.</p>
<p>The final part is the year of collection, which in this example ie either 2004, 2012 or 2014.</p>
<p>Thus, e.g., “frq_BF-09_gamb_2012” means the frequency of the allele within the cohort of <em>Anopheles gambiae</em> mosquitoes from the BF-09 (Haut-Bassins) region collected in 2012.</p>
</div>
<div class="section" id="single-nucleotide-polymorphisms-snps">
<h2>Single nucleotide polymorphisms (SNPs)<a class="headerlink" href="#single-nucleotide-polymorphisms-snps" title="Permalink to this headline">¶</a></h2>
<p>In the last module we learnt about reference genomes and gene annotations. Now one of the cool things we can do once we have an <em>Anopheles gambiae</em> reference genome, is sequence the genomes of wild mosquitoes and compare them to the reference genome. By aligning an individual’s sequencing reads to the reference, we can call its <strong>genotype</strong>.</p>
<p>The genotype is derived from how similar the individual is to the reference genome at each <strong>nucleotide position</strong>.</p>
<p>Where a nucleotide matches the reference genome, it will be called a <strong>reference allele</strong>, where “allele” is just a term for any kind of genetic variant. Where a nucleotide differs from the reference genome, it will be called an <strong>alternative allele</strong>.</p>
<p>Since, there are four possible nucleotides and one of these is the reference, there are always three possible alternative alleles for any given site. Another term for a genetic difference where two individuals differ at a single nucleotide position is <strong>single nucleotide polymorphism</strong> or <strong>SNP</strong>.</p>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><iframe frameborder="0" style="width:100%;height:700px;" src="https://viewer.diagrams.net/?tags=%7B%7D&highlight=000000&edit=_blank&layers=1&nav=1#G1MXxUT74w6lQ7FRppVmza4M4cdS_ZecNw"></iframe></div></div>
</div>
<p>This SNP information is displayed in the first four columns of the <code class="docutils literal notranslate"><span class="pre">snp_allele_frequencies()</span></code> index in the output dataframe. We can see each alternative allele represented by it’s own row in our output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_allele_freqs_df</span><span class="p">[[]]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-bdd1c5e4-f938-4969-a2f5-fa292f83631e">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
    <tr>
      <th>contig</th>
      <th>position</th>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="11" valign="top">2L</th>
      <th rowspan="3" valign="top">2358158</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>M1L</th>
    </tr>
    <tr>
      <th>T</th>
      <th>M1L</th>
    </tr>
    <tr>
      <th>G</th>
      <th>M1V</th>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2358159</th>
      <th rowspan="2" valign="top">T</th>
      <th>A</th>
      <th>M1K</th>
    </tr>
    <tr>
      <th>C</th>
      <th>M1T</th>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <th>...</th>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2431616</th>
      <th rowspan="2" valign="top">G</th>
      <th>C</th>
      <th>*2119S</th>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119L</th>
    </tr>
    <tr>
      <th rowspan="3" valign="top">2431617</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>*2119C</th>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119C</th>
    </tr>
    <tr>
      <th>G</th>
      <th>*2119W</th>
    </tr>
  </tbody>
</table>
<p>209001 rows × 0 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-bdd1c5e4-f938-4969-a2f5-fa292f83631e')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-bdd1c5e4-f938-4969-a2f5-fa292f83631e button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-bdd1c5e4-f938-4969-a2f5-fa292f83631e');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="snp-effects-some-snps-may-be-more-interesting-than-others">
<h2>SNP effects - some SNPs may be more interesting than others<a class="headerlink" href="#snp-effects-some-snps-may-be-more-interesting-than-others" title="Permalink to this headline">¶</a></h2>
<p>SNPs can have different effects depending on what the nucleotide change is and where in the genome it occurs.</p>
<p>In this analysis, we are interested in SNPs that affect protein structure, specifically those which will change the voltage-gated sodium-channel and could affect protein function and therefore insecticide resistance phenotype, so we need to look within the coding sequences (CDS) of the <em>Vgsc</em> gene.</p>
<p>However, not all SNPs which fall in CDSs cause protein changes, because the genetic code has some redundancy, meaning that different nucleotide sequences can encode the same amino acid. If a SNP in a CDS does change the amino acid, it is called a <strong>non-synonymous</strong> (NS) or <strong>missense</strong> SNP, and if the SNP does not change the amino acid, it is called a <strong>synonymous</strong> SNP.</p>
<p>Manually predicting SNP effects is quite involved, but <code class="docutils literal notranslate"><span class="pre">snp_allele_frequencies()</span></code> can predict them for us, let’s look at how this is represented in our output DataFrame.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_effects_df</span> <span class="o">=</span> <span class="n">snp_allele_freqs_df</span><span class="p">[[</span><span class="s2">&quot;effect&quot;</span><span class="p">,</span> <span class="s2">&quot;impact&quot;</span><span class="p">]]</span>
<span class="n">snp_effects_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-17a04152-4465-42c7-a50e-1cab5ad9e17d">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>effect</th>
      <th>impact</th>
    </tr>
    <tr>
      <th>contig</th>
      <th>position</th>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="11" valign="top">2L</th>
      <th rowspan="3" valign="top">2358158</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>M1L</th>
      <td>START_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th>T</th>
      <th>M1L</th>
      <td>START_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th>G</th>
      <th>M1V</th>
      <td>START_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2358159</th>
      <th rowspan="2" valign="top">T</th>
      <th>A</th>
      <th>M1K</th>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
    </tr>
    <tr>
      <th>C</th>
      <th>M1T</th>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2431616</th>
      <th rowspan="2" valign="top">G</th>
      <th>C</th>
      <th>*2119S</th>
      <td>STOP_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119L</th>
      <td>STOP_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">2431617</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>*2119C</th>
      <td>STOP_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119C</th>
      <td>STOP_LOST</td>
      <td>HIGH</td>
    </tr>
    <tr>
      <th>G</th>
      <th>*2119W</th>
      <td>STOP_LOST</td>
      <td>HIGH</td>
    </tr>
  </tbody>
</table>
<p>209001 rows × 2 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-17a04152-4465-42c7-a50e-1cab5ad9e17d')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-17a04152-4465-42c7-a50e-1cab5ad9e17d button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-17a04152-4465-42c7-a50e-1cab5ad9e17d');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>Let’s look specifically at the genomic position where a SNP occurs which causes an insecticide-resistance mutation, also known as “kdr”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_effects_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="s2">&quot;2L&quot;</span><span class="p">,</span> <span class="mi">2_422_652</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-4ccd2bc2-96eb-4959-b973-c62f9924f134">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th>effect</th>
      <th>impact</th>
    </tr>
    <tr>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>L995F</th>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
    </tr>
    <tr>
      <th>T</th>
      <th>L995F</th>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
    </tr>
    <tr>
      <th>G</th>
      <th>L995L</th>
      <td>SYNONYMOUS_CODING</td>
      <td>LOW</td>
    </tr>
  </tbody>
</table>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-4ccd2bc2-96eb-4959-b973-c62f9924f134')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-4ccd2bc2-96eb-4959-b973-c62f9924f134 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-4ccd2bc2-96eb-4959-b973-c62f9924f134');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>For interest, let’s count the number of SNPs we have by their effect.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_allele_freqs_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;max_af &gt; 0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;effect&quot;</span><span class="p">,</span> <span class="s2">&quot;impact&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>effect                 impact  
INTRONIC               MODIFIER    9451
NON_SYNONYMOUS_CODING  MODERATE     121
SPLICE_CORE            HIGH           8
SPLICE_REGION          MODERATE       9
STOP_GAINED            HIGH          14
SYNONYMOUS_CODING      LOW           56
dtype: int64
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="snp-allele-frequencies">
<h2>SNP allele frequencies<a class="headerlink" href="#snp-allele-frequencies" title="Permalink to this headline">¶</a></h2>
<p>Identifying the presence or absence of SNPs in wild caught mosquitoes is interesting, but the real value in generating SNP genotypes from large spatiotemporal collections of mosquitoes comes from the ability to see how groups of samples (cohorts) differ between geographical locations, species, and over time.</p>
<p>One way to compare SNP differences between cohorts is to calculate and compare <strong>SNP allele frequencies</strong> at each position in the genome by dividing the number of times each SNP allele is found in the cohort by the total number of individuals present in the cohort (multiplied by 2 because each individual mosquito is diploid and so carries two genome copies).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%html</span>
<span class="p">&lt;</span><span class="nt">iframe</span> <span class="na">frameborder</span><span class="o">=</span><span class="s">&quot;0&quot;</span> <span class="na">style</span><span class="o">=</span><span class="s">&quot;width:100%;height:700px;&quot;</span> <span class="na">src</span><span class="o">=</span><span class="s">&quot;https://viewer.diagrams.net/?tags=%7B%7D&amp;highlight=0000ff&amp;edit=_blank&amp;layers=1&amp;nav=1&amp;title=allele_freqs.drawio#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D1vbRdL36K4NAeQWOF59o4KR723rZrBnak%26export%3Ddownload&quot;</span><span class="p">&gt;&lt;/</span><span class="nt">iframe</span><span class="p">&gt;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><iframe frameborder="0" style="width:100%;height:700px;" src="https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=allele_freqs.drawio#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D1vbRdL36K4NAeQWOF59o4KR723rZrBnak%26export%3Ddownload"></iframe></div></div>
</div>
<p>Let’s take another look at the allele frequencies we computed above, focusing just on the frequency columns.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_allele_freqs_df</span><span class="p">[</span><span class="n">frequency_columns</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;max_af&#39;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-82aee72d-8e44-4381-af62-2337d00f31bd">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>frq_BF-07_gamb_2004</th>
      <th>max_af</th>
    </tr>
    <tr>
      <th>contig</th>
      <th>position</th>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="11" valign="top">2L</th>
      <th rowspan="3" valign="top">2358158</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>M1L</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>T</th>
      <th>M1L</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>G</th>
      <th>M1V</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2358159</th>
      <th rowspan="2" valign="top">T</th>
      <th>A</th>
      <th>M1K</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>C</th>
      <th>M1T</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th rowspan="2" valign="top">2431616</th>
      <th rowspan="2" valign="top">G</th>
      <th>C</th>
      <th>*2119S</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119L</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">2431617</th>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>*2119C</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>T</th>
      <th>*2119C</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>G</th>
      <th>*2119W</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>209001 rows × 6 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-82aee72d-8e44-4381-af62-2337d00f31bd')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-82aee72d-8e44-4381-af62-2337d00f31bd button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-82aee72d-8e44-4381-af62-2337d00f31bd');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>And let’s inspect the frequencies for SNPs at a specific genomic position of interest.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_allele_freqs_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="s2">&quot;2L&quot;</span><span class="p">,</span> <span class="mi">2_422_652</span><span class="p">),</span> <span class="n">frequency_columns</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>frq_BF-07_gamb_2004</th>
    </tr>
    <tr>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">A</th>
      <th>C</th>
      <th>L995F</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>T</th>
      <th>L995F</th>
      <td>1.0</td>
      <td>0.865854</td>
      <td>0.886792</td>
      <td>1.0</td>
      <td>0.076923</td>
    </tr>
    <tr>
      <th>G</th>
      <th>L995L</th>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-37ee88dc-2ac2-48d2-9553-635ccac3ccfb');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
</div>
<div class="section" id="visualising-snp-allele-frequencies">
<h2>Visualising SNP allele frequencies<a class="headerlink" href="#visualising-snp-allele-frequencies" title="Permalink to this headline">¶</a></h2>
<p>To make our SNP allele frequencies DataFrame easier to interpret, we can filter it down to just non-synonymous SNPs that are at frequency above 5% in at least one of our cohorts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ns_snps_df</span> <span class="o">=</span> <span class="n">snp_allele_freqs_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;effect == &#39;NON_SYNONYMOUS_CODING&#39; and max_af &gt;= 0.05&quot;</span><span class="p">)</span>
<span class="n">ns_snps_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>pass_gamb_colu_arab</th>
      <th>pass_gamb_colu</th>
      <th>pass_arab</th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>frq_BF-07_gamb_2004</th>
      <th>max_af</th>
      <th>effect</th>
      <th>impact</th>
      <th>ref_codon</th>
      <th>alt_codon</th>
      <th>aa_pos</th>
      <th>ref_aa</th>
      <th>alt_aa</th>
    </tr>
    <tr>
      <th>contig</th>
      <th>position</th>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="13" valign="top">2L</th>
      <th rowspan="2" valign="top">2391228</th>
      <th rowspan="2" valign="top">G</th>
      <th>C</th>
      <th>V402L</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.067073</td>
      <td>0.028302</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.067073</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>Gta</td>
      <td>Cta</td>
      <td>402.0</td>
      <td>V</td>
      <td>L</td>
    </tr>
    <tr>
      <th>T</th>
      <th>V402L</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.054878</td>
      <td>0.084906</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.084906</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>Gta</td>
      <td>Tta</td>
      <td>402.0</td>
      <td>V</td>
      <td>L</td>
    </tr>
    <tr>
      <th>2416980</th>
      <th>C</th>
      <th>T</th>
      <th>T791M</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.163265</td>
      <td>0.018293</td>
      <td>0.000000</td>
      <td>0.239130</td>
      <td>0.000000</td>
      <td>0.239130</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>aCg</td>
      <td>aTg</td>
      <td>791.0</td>
      <td>T</td>
      <td>M</td>
    </tr>
    <tr>
      <th>2422652</th>
      <th>A</th>
      <th>T</th>
      <th>L995F</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>1.000000</td>
      <td>0.865854</td>
      <td>0.886792</td>
      <td>1.000000</td>
      <td>0.076923</td>
      <td>1.000000</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>ttA</td>
      <td>ttT</td>
      <td>995.0</td>
      <td>L</td>
      <td>F</td>
    </tr>
    <tr>
      <th>2429617</th>
      <th>T</th>
      <th>C</th>
      <th>I1527T</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.121951</td>
      <td>0.113208</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.121951</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>aTt</td>
      <td>aCt</td>
      <td>1527.0</td>
      <td>I</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2429745</th>
      <th>A</th>
      <th>T</th>
      <th>N1570Y</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.209184</td>
      <td>0.250000</td>
      <td>0.320755</td>
      <td>0.141304</td>
      <td>0.038462</td>
      <td>0.320755</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>Aat</td>
      <td>Tat</td>
      <td>1570.0</td>
      <td>N</td>
      <td>Y</td>
    </tr>
    <tr>
      <th>2429897</th>
      <th>A</th>
      <th>G</th>
      <th>E1597G</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.066327</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.032609</td>
      <td>0.000000</td>
      <td>0.066327</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>gAa</td>
      <td>gGa</td>
      <td>1597.0</td>
      <td>E</td>
      <td>G</td>
    </tr>
    <tr>
      <th>2429915</th>
      <th>A</th>
      <th>C</th>
      <th>K1603T</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.054878</td>
      <td>0.056604</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.056604</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>aAg</td>
      <td>aCg</td>
      <td>1603.0</td>
      <td>K</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2430424</th>
      <th>G</th>
      <th>T</th>
      <th>A1746S</th>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>0.153061</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.239130</td>
      <td>0.000000</td>
      <td>0.239130</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>Gcc</td>
      <td>Tcc</td>
      <td>1746.0</td>
      <td>A</td>
      <td>S</td>
    </tr>
    <tr>
      <th>2430863</th>
      <th>T</th>
      <th>C</th>
      <th>I1868T</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.250000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.206522</td>
      <td>0.000000</td>
      <td>0.250000</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>aTa</td>
      <td>aCa</td>
      <td>1868.0</td>
      <td>I</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2430880</th>
      <th>C</th>
      <th>T</th>
      <th>P1874S</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.213415</td>
      <td>0.169811</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.213415</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>Cct</td>
      <td>Tct</td>
      <td>1874.0</td>
      <td>P</td>
      <td>S</td>
    </tr>
    <tr>
      <th>2430881</th>
      <th>C</th>
      <th>T</th>
      <th>P1874L</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.224490</td>
      <td>0.073171</td>
      <td>0.056604</td>
      <td>0.260870</td>
      <td>0.000000</td>
      <td>0.260870</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>cCt</td>
      <td>cTt</td>
      <td>1874.0</td>
      <td>P</td>
      <td>L</td>
    </tr>
    <tr>
      <th>2431061</th>
      <th>C</th>
      <th>T</th>
      <th>A1934V</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.121951</td>
      <td>0.132075</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.132075</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>gCt</td>
      <td>gTt</td>
      <td>1934.0</td>
      <td>A</td>
      <td>V</td>
    </tr>
  </tbody>
</table>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-b50236b9-840f-4702-a5eb-cb9cd69aeeb2');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>To make things even clearer, we have included a heatmap plotting function to style our filtered DataFrame, called <a class="reference external" href="https://malariagen.github.io/vector-data/ag3/api.html#plot-frequencies-heatmap"><code class="docutils literal notranslate"><span class="pre">plot_frequencies_heatmap()</span></code></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_frequencies_heatmap</span><span class="p">(</span><span class="n">ns_snps_df</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><html>
<head><meta charset="utf-8" /></head>
<body>
    <div>            <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG"></script><script type="text/javascript">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}</script>                <script type="text/javascript">window.PlotlyConfig = {MathJaxConfig: 'local'};</script>
        <script src="https://cdn.plot.ly/plotly-2.8.3.min.js"></script>                <div id="37d951e7-ec15-4ad6-b5ba-f109a526b881" class="plotly-graph-div" style="height:525px; width:600px;"></div>            <script type="text/javascript">                                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById("37d951e7-ec15-4ad6-b5ba-f109a526b881")) {                    Plotly.newPlot(                        "37d951e7-ec15-4ad6-b5ba-f109a526b881",                        [{"coloraxis":"coloraxis","name":"0","texttemplate":"%{z:.0%}","x":["BF-09_gamb_2012","BF-09_colu_2012","BF-09_colu_2014","BF-09_gamb_2014","BF-07_gamb_2004"],"y":["2L, 2391228, G, C, V402L","2L, 2391228, G, T, V402L","2L, 2416980, C, T, T791M","2L, 2422652, A, T, L995F","2L, 2429617, T, C, I1527T","2L, 2429745, A, T, N1570Y","2L, 2429897, A, G, E1597G","2L, 2429915, A, C, K1603T","2L, 2430424, G, T, A1746S","2L, 2430863, T, C, I1868T","2L, 2430880, C, T, P1874S","2L, 2430881, C, T, P1874L","2L, 2431061, C, T, A1934V"],"z":[[0.0,0.06707317073170732,0.02830188679245283,0.0,0.0],[0.0,0.054878048780487805,0.08490566037735849,0.0,0.0],[0.16326530612244897,0.018292682926829267,0.0,0.2391304347826087,0.0],[1.0,0.8658536585365854,0.8867924528301887,1.0,0.07692307692307693],[0.0,0.12195121951219512,0.11320754716981132,0.0,0.0],[0.20918367346938777,0.25,0.32075471698113206,0.14130434782608695,0.038461538461538464],[0.0663265306122449,0.0,0.0,0.03260869565217391,0.0],[0.0,0.054878048780487805,0.05660377358490566,0.0,0.0],[0.15306122448979592,0.0,0.0,0.2391304347826087,0.0],[0.25,0.0,0.0,0.20652173913043478,0.0],[0.0,0.21341463414634146,0.16981132075471697,0.0,0.0],[0.22448979591836735,0.07317073170731707,0.05660377358490566,0.2608695652173913,0.0],[0.0,0.12195121951219512,0.1320754716981132,0.0,0.0]],"type":"heatmap","xaxis":"x","yaxis":"y","hovertemplate":"x: %{x}<br>y: %{y}<br>color: %{z}<extra></extra>"}],                        {"template":{"data":{"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"choropleth":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"choropleth"}],"contour":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"contour"}],"contourcarpet":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"contourcarpet"}],"heatmap":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"heatmap"}],"heatmapgl":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"heatmapgl"}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"histogram2d":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"histogram2d"}],"histogram2dcontour":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"histogram2dcontour"}],"mesh3d":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"mesh3d"}],"parcoords":[{"line":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"parcoords"}],"pie":[{"automargin":true,"type":"pie"}],"scatter":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatter"}],"scatter3d":[{"line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatter3d"}],"scattercarpet":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattercarpet"}],"scattergeo":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattergeo"}],"scattergl":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattergl"}],"scattermapbox":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattermapbox"}],"scatterpolar":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterpolar"}],"scatterpolargl":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterpolargl"}],"scatterternary":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterternary"}],"surface":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"surface"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}]},"layout":{"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"autotypenumbers":"strict","coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]],"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]},"colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"geo":{"bgcolor":"white","lakecolor":"white","landcolor":"#E5ECF6","showlakes":true,"showland":true,"subunitcolor":"white"},"hoverlabel":{"align":"left"},"hovermode":"closest","mapbox":{"style":"light"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"bgcolor":"#E5ECF6","radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"ternary":{"aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"bgcolor":"#E5ECF6","caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"title":{"x":0.05},"xaxis":{"automargin":true,"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","zerolinewidth":2},"yaxis":{"automargin":true,"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","zerolinewidth":2}}},"xaxis":{"anchor":"y","domain":[0.0,1.0],"side":"top","tickangle":-90,"title":{"text":"cohorts"}},"yaxis":{"anchor":"x","domain":[0.0,1.0],"autorange":"reversed"},"coloraxis":{"colorscale":[[0.0,"rgb(255,245,240)"],[0.125,"rgb(254,224,210)"],[0.25,"rgb(252,187,161)"],[0.375,"rgb(252,146,114)"],[0.5,"rgb(251,106,74)"],[0.625,"rgb(239,59,44)"],[0.75,"rgb(203,24,29)"],[0.875,"rgb(165,15,21)"],[1.0,"rgb(103,0,13)"]],"cmin":0,"cmax":1,"colorbar":{"title":{"text":"frequency"},"tickvals":[0,0.2,0.4,0.6,0.8,1.0],"ticktext":["0%","20%","40%","60%","80%","100%"]}},"margin":{"t":60},"width":600},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('37d951e7-ec15-4ad6-b5ba-f109a526b881');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };                            </script>        </div>
</body>
</html></div></div>
</div>
</div>
<div class="section" id="amino-acid-substitution-frequencies">
<h2>Amino acid substitution frequencies<a class="headerlink" href="#amino-acid-substitution-frequencies" title="Permalink to this headline">¶</a></h2>
<p>You might have noticed that there are two rows with <code class="docutils literal notranslate"><span class="pre">V402L</span></code> in our previous heatmap plot. This is because in Burkina Faso, we find two different alternative alleles at the same genomic position, both causing the same amino acid substitution (valine to leucine).</p>
<p>If we are just interested in amino acid change frequencies, for example, when looking at potential insecticide resistance conferring mutations, we might want to combine the frequencies of the two alleles which cause V402L. In this case, we can use the <code class="docutils literal notranslate"><span class="pre">aa_allele_frequencies()</span></code> function in exactly the same way as we used <code class="docutils literal notranslate"><span class="pre">snp_allele_frequencies()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aa_allele_freqs_df</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">aa_allele_frequencies</span><span class="p">(</span>
    <span class="n">transcript</span><span class="o">=</span><span class="n">transcript</span><span class="p">,</span> 
    <span class="n">cohorts</span><span class="o">=</span><span class="n">cohorts</span><span class="p">,</span> 
    <span class="n">sample_sets</span><span class="o">=</span><span class="n">sample_sets</span>
<span class="p">)</span>
<span class="n">aa_allele_freqs_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-2bb8a67f-f954-4972-bdbc-769b1d462936">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>frq_BF-07_gamb_2004</th>
      <th>max_af</th>
    </tr>
    <tr>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A32V</th>
      <td>0.000000</td>
      <td>0.006098</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.006098</td>
    </tr>
    <tr>
      <th>G54C</th>
      <td>0.000000</td>
      <td>0.018293</td>
      <td>0.0</td>
      <td>0.010870</td>
      <td>0.000000</td>
      <td>0.018293</td>
    </tr>
    <tr>
      <th>P55L</th>
      <td>0.005102</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.010870</td>
      <td>0.000000</td>
      <td>0.010870</td>
    </tr>
    <tr>
      <th>P59T</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.021739</td>
      <td>0.000000</td>
      <td>0.021739</td>
    </tr>
    <tr>
      <th>G73D</th>
      <td>0.000000</td>
      <td>0.006098</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.006098</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>A2023G</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.038462</td>
      <td>0.038462</td>
    </tr>
    <tr>
      <th>S2037R</th>
      <td>0.005102</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.005102</td>
    </tr>
    <tr>
      <th>I2053V</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.038462</td>
      <td>0.038462</td>
    </tr>
    <tr>
      <th>G2055V</th>
      <td>0.005102</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.005102</td>
    </tr>
    <tr>
      <th>V2118A</th>
      <td>0.005102</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.005102</td>
    </tr>
  </tbody>
</table>
<p>134 rows × 6 columns</p>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-2bb8a67f-f954-4972-bdbc-769b1d462936')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-2bb8a67f-f954-4972-bdbc-769b1d462936 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-2bb8a67f-f954-4972-bdbc-769b1d462936');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>Let’s filter it again to just amino acid changes greater than 5% in at least one cohort. We don’t need to filter for non-synonymous mutations this time as this function has already done that for us.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aa_filt_df</span> <span class="o">=</span> <span class="n">aa_allele_freqs_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;max_af &gt;= 0.05&quot;</span><span class="p">)</span>
<span class="n">aa_filt_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
  <div id="df-88a4a758-bb32-4762-b31d-203365b338d0">
    <div class="colab-df-container">
      <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>frq_BF-07_gamb_2004</th>
      <th>max_af</th>
    </tr>
    <tr>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>V402L</th>
      <td>0.000000</td>
      <td>0.121951</td>
      <td>0.113208</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.121951</td>
    </tr>
    <tr>
      <th>T791M</th>
      <td>0.163265</td>
      <td>0.018293</td>
      <td>0.000000</td>
      <td>0.239130</td>
      <td>0.000000</td>
      <td>0.239130</td>
    </tr>
    <tr>
      <th>L995F</th>
      <td>1.000000</td>
      <td>0.865854</td>
      <td>0.886792</td>
      <td>1.000000</td>
      <td>0.076923</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>I1527T</th>
      <td>0.000000</td>
      <td>0.121951</td>
      <td>0.113208</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.121951</td>
    </tr>
    <tr>
      <th>N1570Y</th>
      <td>0.209184</td>
      <td>0.250000</td>
      <td>0.320755</td>
      <td>0.141304</td>
      <td>0.038462</td>
      <td>0.320755</td>
    </tr>
    <tr>
      <th>E1597G</th>
      <td>0.066327</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.032609</td>
      <td>0.000000</td>
      <td>0.066327</td>
    </tr>
    <tr>
      <th>K1603T</th>
      <td>0.000000</td>
      <td>0.054878</td>
      <td>0.056604</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.056604</td>
    </tr>
    <tr>
      <th>A1746S</th>
      <td>0.153061</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.239130</td>
      <td>0.000000</td>
      <td>0.239130</td>
    </tr>
    <tr>
      <th>I1868T</th>
      <td>0.250000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.206522</td>
      <td>0.000000</td>
      <td>0.250000</td>
    </tr>
    <tr>
      <th>P1874L</th>
      <td>0.224490</td>
      <td>0.073171</td>
      <td>0.056604</td>
      <td>0.260870</td>
      <td>0.000000</td>
      <td>0.260870</td>
    </tr>
    <tr>
      <th>P1874S</th>
      <td>0.000000</td>
      <td>0.213415</td>
      <td>0.169811</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.213415</td>
    </tr>
    <tr>
      <th>A1934V</th>
      <td>0.000000</td>
      <td>0.121951</td>
      <td>0.132075</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.132075</td>
    </tr>
  </tbody>
</table>
</div>
      <button class="colab-df-convert" onclick="convertToInteractive('df-88a4a758-bb32-4762-b31d-203365b338d0')"
              title="Convert this dataframe to an interactive table."
              style="display:none;">

  <svg xmlns="http://www.w3.org/2000/svg" height="24px"viewBox="0 0 24 24"
       width="24px">
    <path d="M0 0h24v24H0V0z" fill="none"/>
    <path d="M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z"/><path d="M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z"/>
  </svg>
      </button>

  <style>
    .colab-df-container {
      display:flex;
      flex-wrap:wrap;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>

      <script>
        const buttonEl =
          document.querySelector('#df-88a4a758-bb32-4762-b31d-203365b338d0 button.colab-df-convert');
        buttonEl.style.display =
          google.colab.kernel.accessAllowed ? 'block' : 'none';

        async function convertToInteractive(key) {
          const element = document.querySelector('#df-88a4a758-bb32-4762-b31d-203365b338d0');
          const dataTable =
            await google.colab.kernel.invokeFunction('convertToInteractive',
                                                     [key], {});
          if (!dataTable) return;

          const docLinkHtml = 'Like what you see? Visit the ' +
            '<a target="_blank" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'
            + ' to learn more about interactive tables.';
          element.innerHTML = '';
          dataTable['output_type'] = 'display_data';
          await google.colab.output.renderOutput(dataTable, element);
          const docLink = document.createElement('div');
          docLink.innerHTML = docLinkHtml;
          element.appendChild(docLink);
        }
      </script>
    </div>
  </div>
</div></div>
</div>
<p>Now we can visualise these frequencies the same way we did before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_frequencies_heatmap</span><span class="p">(</span><span class="n">aa_filt_df</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><html>
<head><meta charset="utf-8" /></head>
<body>
    <div>            <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG"></script><script type="text/javascript">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}</script>                <script type="text/javascript">window.PlotlyConfig = {MathJaxConfig: 'local'};</script>
        <script src="https://cdn.plot.ly/plotly-2.8.3.min.js"></script>                <div id="2f468535-afef-42e0-a924-93e51f71826f" class="plotly-graph-div" style="height:525px; width:600px;"></div>            <script type="text/javascript">                                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById("2f468535-afef-42e0-a924-93e51f71826f")) {                    Plotly.newPlot(                        "2f468535-afef-42e0-a924-93e51f71826f",                        [{"coloraxis":"coloraxis","name":"0","texttemplate":"%{z:.0%}","x":["BF-09_gamb_2012","BF-09_colu_2012","BF-09_colu_2014","BF-09_gamb_2014","BF-07_gamb_2004"],"y":["V402L","T791M","L995F","I1527T","N1570Y","E1597G","K1603T","A1746S","I1868T","P1874L","P1874S","A1934V"],"z":[[0.0,0.12195121951219512,0.11320754716981132,0.0,0.0],[0.16326530612244897,0.018292682926829267,0.0,0.2391304347826087,0.0],[1.0,0.8658536585365854,0.8867924528301887,1.0,0.07692307692307693],[0.0,0.12195121951219512,0.11320754716981132,0.0,0.0],[0.20918367346938777,0.25,0.32075471698113206,0.14130434782608695,0.038461538461538464],[0.0663265306122449,0.0,0.0,0.03260869565217391,0.0],[0.0,0.054878048780487805,0.05660377358490566,0.0,0.0],[0.15306122448979592,0.0,0.0,0.2391304347826087,0.0],[0.25,0.0,0.0,0.20652173913043478,0.0],[0.22448979591836735,0.07317073170731707,0.05660377358490566,0.2608695652173913,0.0],[0.0,0.21341463414634146,0.16981132075471697,0.0,0.0],[0.0,0.12195121951219512,0.1320754716981132,0.0,0.0]],"type":"heatmap","xaxis":"x","yaxis":"y","hovertemplate":"x: %{x}<br>y: %{y}<br>color: %{z}<extra></extra>"}],                        {"template":{"data":{"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"choropleth":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"choropleth"}],"contour":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"contour"}],"contourcarpet":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"contourcarpet"}],"heatmap":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"heatmap"}],"heatmapgl":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"heatmapgl"}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"histogram2d":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"histogram2d"}],"histogram2dcontour":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"histogram2dcontour"}],"mesh3d":[{"colorbar":{"outlinewidth":0,"ticks":""},"type":"mesh3d"}],"parcoords":[{"line":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"parcoords"}],"pie":[{"automargin":true,"type":"pie"}],"scatter":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatter"}],"scatter3d":[{"line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatter3d"}],"scattercarpet":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattercarpet"}],"scattergeo":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattergeo"}],"scattergl":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattergl"}],"scattermapbox":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scattermapbox"}],"scatterpolar":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterpolar"}],"scatterpolargl":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterpolargl"}],"scatterternary":[{"marker":{"colorbar":{"outlinewidth":0,"ticks":""}},"type":"scatterternary"}],"surface":[{"colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"type":"surface"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}]},"layout":{"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"autotypenumbers":"strict","coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]],"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]},"colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"geo":{"bgcolor":"white","lakecolor":"white","landcolor":"#E5ECF6","showlakes":true,"showland":true,"subunitcolor":"white"},"hoverlabel":{"align":"left"},"hovermode":"closest","mapbox":{"style":"light"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"bgcolor":"#E5ECF6","radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","gridwidth":2,"linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white"}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"ternary":{"aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"bgcolor":"#E5ECF6","caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"title":{"x":0.05},"xaxis":{"automargin":true,"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","zerolinewidth":2},"yaxis":{"automargin":true,"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","zerolinewidth":2}}},"xaxis":{"anchor":"y","domain":[0.0,1.0],"side":"top","tickangle":-90,"title":{"text":"cohorts"}},"yaxis":{"anchor":"x","domain":[0.0,1.0],"autorange":"reversed"},"coloraxis":{"colorscale":[[0.0,"rgb(255,245,240)"],[0.125,"rgb(254,224,210)"],[0.25,"rgb(252,187,161)"],[0.375,"rgb(252,146,114)"],[0.5,"rgb(251,106,74)"],[0.625,"rgb(239,59,44)"],[0.75,"rgb(203,24,29)"],[0.875,"rgb(165,15,21)"],[1.0,"rgb(103,0,13)"]],"cmin":0,"cmax":1,"colorbar":{"title":{"text":"frequency"},"tickvals":[0,0.2,0.4,0.6,0.8,1.0],"ticktext":["0%","20%","40%","60%","80%","100%"]}},"margin":{"t":60},"width":600},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('2f468535-afef-42e0-a924-93e51f71826f');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };                            </script>        </div>
</body>
</html></div></div>
</div>
</div>
<div class="section" id="well-done">
<h2>Well done!<a class="headerlink" href="#well-done" title="Permalink to this headline">¶</a></h2>
<p>In this module we have learnt how to analyse SNP mutations in the target of pyrethroid insecticides, the voltage-gated sodium-channel. We have calculated the allele frequencies of the SNPs in cohorts of mosquitoes and learnt how to filter and plot them for ease of interpretation.</p>
</div>
<div class="section" id="practical-exercises">
<h2>Practical exercises<a class="headerlink" href="#practical-exercises" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>English<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Open this notebook in Google Colab and run it for yourself from top to bottom. Hint: click the rocket icon at the top of the page, then select “Colab” from the drop-down menu.</p></li>
<li><p>Looking at the heatmap output (either amino acid or SNP), can you spot a relationship between the <code class="docutils literal notranslate"><span class="pre">V402L</span></code> and <code class="docutils literal notranslate"><span class="pre">L995F</span></code> frequencies? If so, what is it?</p></li>
<li><p>Re-run the whole analysis but using the Ghanaian sample set. Hint: Try <code class="docutils literal notranslate"><span class="pre">sample_sets</span> <span class="pre">=</span> <span class="pre">&quot;AG1000G-GH&quot;</span></code>. Or any other samples of interest.</p></li>
<li><p>What are the cohorts for this new sample set? Hint: see <code class="docutils literal notranslate"><span class="pre">frequency_columns</span></code>.</p></li>
<li><p>Above, we looked at the <em>kdr</em> “West” SNP position, compare and contrast this with the <em>kdr</em> “East” SNP position. What is the amino acid change for <em>kdr</em> “East”? Hint: The position is 2422651.</p></li>
<li><p>For the Ghanaian sample set, add the a y_label which says “aa change” to the amino acid frequency heatmap. Hint: you can view all the function’s parameters with <code class="docutils literal notranslate"><span class="pre">ag3.aa_allele_frequencies?</span></code>.</p></li>
<li><p>Remove the colorbar from the same heatmap. Hint: <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p>Is the same relationship between <code class="docutils literal notranslate"><span class="pre">V402L</span></code> and <code class="docutils literal notranslate"><span class="pre">L995F</span></code> frequencies present in Ghana? What might be an evolutionary interpretation of this relationship?</p></li>
</ol>
<p><strong>When you’ve had enough, create a link to your notebook, and share it with someone. If you’re attending a training workshop, paste the link into the workshop slack channel.</strong></p>
</div>
<div class="section" id="id2">
<h3>Français<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>Coming soon.</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./workshop-1"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="module-3-genome.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Module 3 - Navigating the <em>Anopheles gambiae</em> reference genome</p>
        </div>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By MalariaGEN and PAMCA<br/>
    
        &copy; Copyright 2022.<br/>
      <div class="extra_footer">
        <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a> This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.

      </div>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>